---
title: "Analysis of the CBMC data"
author: "[S.M.Pagnotta](https://www.researchgate.net/profile/Stefano_Pagnotta) - Dept. of Science and Technology, Universita' degli Studi del Sannio (Benevento, Italy)</br> [D.Risso](https://www.researchgate.net/profile/Davide_Risso) - Dept. of Statistical Sciences, Universita' degli Studi di Padova (Padova, Italy)"
date:  '`r format(Sys.time(), "%d %B, %Y")`'
output: 
  html_document: 
    toc: yes
    toc_float: yes
    df_print: paged
fontsize: 11pt
editor_options: 
  chunk_output_type: console
---

```{r knitr, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.height = 4.5, fig.width =  11, fig.align = "center")
```

# Install and load `awst`

```{r install_awst, eval=FALSE, echo=TRUE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("drisso/awst")
```

```{r setup, message=FALSE}
library(readr)
library(awst)
library(EDASeq)
library(Rtsne)
library(umap)
library(dendextend)
```

# Data import and cleaning

The data are available on GEO with the following accession number.

* [GSE100866](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE100866)
    + [GSE100866_PBMC_vs_flow_10X-ADT_umi.csv.gz](https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE100866&format=file&file=GSE100866%5FPBMC%5Fvs%5Fflow%5F10X%2DADT%5Fumi%2Ecsv%2Egz)
    + [GSE100866_PBMC_vs_flow_10X-RNA_umi.csv.gz](https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE100866&format=file&file=GSE100866%5FPBMC%5Fvs%5Fflow%5F10X%2DRNA%5Fumi%2Ecsv%2Egz)

In the following chunks of code, we arrange a data-frame where each of the cell are annotated, and the annotations are coded as colors.

We mostly follow [the original article](https://www.nature.com/articles/nmeth.4380) for data preprocessing.

## ADT and annotation

```{r annotation}
if(!file.exists("annotation.RData")) {
  positive.col <-  "red"
  negative.col <- "gray"
  ttable <- read_csv("https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE100866&format=file&file=GSE100866_PBMC_vs_flow_10X-ADT_umi.csv.gz")
  ttable <- as.data.frame(ttable)
  rownames(ttable) <- ttable[,1]
  ttable <- ttable[,-1]
  ttable <- 1 + t(ttable)
  tmp <- exp(log(apply(ttable, 1, prod))/ncol(ttable))
  ttable <- log(ttable/tmp)
  ttable <- 6 * (pnorm(ttable) - 0.5)
  annotation.df <- as.data.frame(ttable)
  annotation.df$cell <- rownames(annotation.df)
  nBins <- 9
  bbreaks <- seq(-3, 3, length.out = nBins + 1)
  levels_cols <- c("green4", "green3", "green2", "green", "gray75", "red", "red2", "red3", "red4")
  j <- 0
  while(j < ncol(ttable)) {
    j <- j + 1
    annotation.df$tmp <- cut(ttable[, j], breaks = bbreaks, include.lowest = TRUE)
    levels(annotation.df$tmp) <- levels_cols
    colnames(annotation.df)[ncol(annotation.df)] <- paste0(colnames(ttable)[j], ".col")
  }
  save(annotation.df, file = "annotation.RData")
} else {
  load("annotation.RData")
}
```


# Single-cell RNA-seq

Before applying normalization and AWST, we filtered out cells that did not met the following criteria: 

* remove the genes detected in no cells.

* filter out cells not having at least 500 observed features.

* filter out cells not having at least 20 different intensities across the features.

```{r Level3}
if(!file.exists("Level3.RData")) {
ddata <- read_csv("https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE100866&format=file&file=GSE100866_PBMC_vs_flow_10X-RNA_umi.csv.gz")

#restrict the features to the HUMAN ones
ddata <- ddata[grep("^HUMAN_", ddata$X1),]
ddata <- as.data.frame(ddata)
rownames(ddata) <- gsub("HUMAN_", "", ddata$X1)
ddata <- as.matrix(ddata[,-1])
save(ddata, file = "Level3.RData")
} else {
  load("Level3.RData")
}
```

## Apply AWST

```{r expression.RData}
if(!file.exists("expression.RData")) {
  ddata <- t(ddata)
  ###
  no_of_detected_gene_per_sample <- rowSums(ddata > 0) 
  fivenum(no_of_detected_gene_per_sample)#  10      638      739      873         4833 
  # restrict the collection of cells to those cells having at least 500 observed features
  sum(no_of_detected_gene_per_sample > 500) # 7613
  dim(ddata <- ddata[no_of_detected_gene_per_sample > 500,])#[1]  7613 17014
  ###
  no_of_different_intensities <- apply(ddata, 1, function(x) length(table(x)))
  fivenum(no_of_different_intensities)#  9     17   20     24    105 
  # restrict the collection of cells to those cells having at least 20 different intensities across the featutures
  dim(ddata <- ddata[which(no_of_different_intensities >= 20), ])#[1] 4123 17014
  ###
  # apply the full quantile normalization 
  normCounts <- EDASeq::betweenLaneNormalization(t(ddata), which = "full", round = FALSE)
  save(normCounts, file = "normCounts.RData")
  ###
  # apply the AWS-transformation 
  library(awst)
  dim(exprData <- awst(normCounts, poscount = TRUE, full_quantile = TRUE)) #[1]  4123 17014
  dim(exprData <- gene_filter(exprData, nBins = 30, heterogeneity_threshold = 0.05)) #[1] 4123  330
  save(exprData, file = "expression.RData")
} else {
  load("expression.RData")
}
```

## Clustering and dimensionality reduction

```{r working matrices}
if(!file.exists("expression_umap_2d.RData")) {
nrow_exprData <- nrow(exprData)
ncol_exprData <- ncol(exprData)
ddist <- dist(exprData)
save(ddist, nrow_exprData, ncol_exprData, file = "expression_dist.RData")

hhc <- hclust(ddist, method = "ward.D2")
save(hhc, nrow_exprData, ncol_exprData, file = "expression_dist_hclust.RData")

pprcomp <- prcomp(exprData)
pprcomp$x <- pprcomp$x[, 1:10]
pprcomp$rotation <- pprcomp$rotation[, 1:10]
save(pprcomp, file = "expression_prcomp.RData")

set.seed(2019) # needed to get the figure in the paper
ans_Rtsne <- Rtsne(exprData, pca = FALSE, perplexity = 250) # Run TSNE
save(ans_Rtsne, file = "expression_Rtsne_2d.RData")

set.seed(2019) # needed to get the figure in the paper
ans_umap <- umap(exprData)
save(ans_umap, file = "expression_umap_2d.RData")
}
```

```{r figure_setup, eval=TRUE}
load("annotation.RData")
load("expression_dist_hclust.RData")
annotation.df <- annotation.df[hhc$labels,]
###############
save_plots <- FALSE
png_width_large <- 1500
png_height_large <- 750
png_width_small <- width_png <- 700
png_height_small <- 700
png_res <- 1/300
###################
color.bar2 <- function(x_pos, y_pos, lut, min, max=-min, nticks=11, ticks=seq(min, max, len=nticks), title='', values = NULL) {
    scale = (max-min)/length(lut)*0.3
    for (i in 1:length(lut)) {
      y_low  <- (i-1)*scale + min + y_pos
    	y_high <-  y_low + scale
    	rect(x_pos,y_low,x_pos+.05,y_high, col=lut[i], border=NA)
    	text(x_pos+.05, (y_low + y_high)/2, values[i], adj = -0.1)
    }	
}
vvalues <- c("-3.0", "-2.0", "-1.3", "-0.6", " 0.0", " 0.6", " 1.3", " 2.0", " 3.0")
ffill2 <- names(table(annotation.df$CD3.col))
```

# Main Clustering

```{r _main_dendrogram.pdf, echo=FALSE, eval=TRUE, fig.width=30, fig.height=13, echo=TRUE, eval=TRUE}
clustering.prefix <- "CBMC"; short.prefix <- "CBMC"
clustering.df <- data.frame(cell = annotation.df$cell)
rownames(clustering.df) <- clustering.df$cell

############
mmain  <-  paste0("CBMC study (", nrow_exprData, " cells/", ncol_exprData, " genes)")
if(save_plots) {
  mmain <- ""; png("main_dendrogram.png", width= png_width_large, height= png_height_large, res = png_res)
}
hhc$height <- hhc$height/max(hhc$height)

plot(hhc, hang = -1, labels = FALSE, xlab = "", sub = "", main = mmain)
###
wwhere <- 10
hh <- mean(c(hhc$height[length(hhc$height)-wwhere+2], hhc$height[length(hhc$height)-wwhere+1]))

tmp <- tmp_ <- as.factor(cutree(hhc, k = wwhere))

levels(tmp) <- c( "1", "1", "3", "4", "5", "6", "6", "8","4", "5")
wwhere <- length(unique(levels(tmp)))
clusteringWhere <- paste0(clustering.prefix, wwhere)
clusteringWhere.col <- paste0(clusteringWhere, ".col")
assign(clusteringWhere.col, tmp)
if(wwhere < 10) levels(tmp) <- paste0(short.prefix, wwhere, 1:wwhere) else levels(tmp) <- paste0(short.prefix, wwhere, c(paste(1:9), letters[1:(wwhere-9)]))
assign(clusteringWhere, tmp)
levels(tmp) <- c("black", "red", "green3", "blue", "cyan", "magenta")

assign(clusteringWhere.col, tmp)
tt <- table(get(clusteringWhere), get(clusteringWhere.col))
colorCode <- colnames(tt)[apply(tt, 1, which.max)]
names(colorCode) <- rownames(tt)
assign(paste0(clusteringWhere, ".colorCode"), colorCode)
clust.colorCode <- colorCode
clustering.df$tmp <- get(clusteringWhere)
clustering.df$tmp.explanatory <- clustering.df$tmp
clustering.df$tmp.col <- get(clusteringWhere.col)
ncol_ <- ncol(clustering.df)
colnames(clustering.df)[(ncol_-2):ncol_] <- c(clusteringWhere, paste0(clusteringWhere, ".explanatory"), clusteringWhere.col)
levels(clustering.df[, paste0(clusteringWhere, ".explanatory")]) <- c("CBMC1 - T Cell", "CBMC2 - B Cell", "CBMC3 - unclear", "CBMC4 - Monocyte", "CBMC5 - myeloid DC", "CBMC6 - plasmacytoid DC")

annotation.col <- annotation.df[, grep(".col", colnames(annotation.df))]
colnames(annotation.col) <- gsub(".col", "", colnames(annotation.col))
annotation.col <- annotation.col[, rev(c( "CD19", "CD3",  "CD11c",  "CD14","CD4","CD8", "CD2","CD57"))]
annotation.col$CBMC <- clustering.df[, ncol(clustering.df)]

colored_bars(colors = annotation.col, dend = as.dendrogram(hhc), y_scale = 0.17, y_shift = 0.015)
save(clustering.df, clust.colorCode, file = "clustering.RData")

tt <- table(clustering.df[, ncol(clustering.df)-1])
pct <- paste0(round(100*tt/sum(tt), 1), "%")
llegend <- paste(names(tt), " (", tt, "; ", pct, ")", sep = "")
tt <- table(clustering.df[, ncol(clustering.df)-1], clustering.df[, ncol(clustering.df)])
ffill <- colnames(tt)[apply(tt, 1, which.max)]
legend(700, .99, legend=llegend, fill = ffill, y.intersp = 1, box.col = "white", border = "white", title = "CBMC", title.adj = 0)

color.bar3 <- function(x_pos, y_pos, lut, min, max=-min, nticks=11, ticks=seq(min, max, len=nticks), title='', values = NULL) {
    scale = (max-min)/length(lut)*0.3
    for (i in 1:length(lut)) {
      y_low  <- (i-1)*scale + y_pos
    	y_high <-  y_low + scale
    	rect(x_pos,y_low,x_pos+90,y_high, col=lut[i], border=NA)
    	text(x_pos+.05, (y_low + y_high)/2, values[i], adj = -2)
    }	
}
color.bar3(4000, 0.65, ffill2, -0.6, values = vvalues)
text(4045, 0.63, "Marker's level")
text(1, 1, "(a)")
```

# PCA (AWST) | ADT

```{r _pprcomp_study.pdf, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis'}
x_legend <- -0.2; y_legend <- 2.63
x_text <- -2.3; y_text <- 2.
load("expression_prcomp.RData")
pprcomp$x <- scale(pprcomp$x)
mmain = paste0("Principal components analysis (", nrow_exprData, " cells/", ncol_exprData, " features)")
if(save_plots) png("prcomp_CITE6.png", width= png_width_small, height= png_height_small, res = png_res)
plot(pprcomp$x, col = clustering.df$CBMC6.col, main = mmain, 
     xlab = "first principal component", ylab = "secondo principal component", pch = 19)
legend(x_legend, y_legend, legend=llegend, fill = ffill, y.intersp = 1, box.col = "white", border = "white", title = "CBMC", title.adj = 0)
text(x_text, y_text, "(b)")
###############################

mmain <-  "principal component analysis - CD3"#  "prcomp (AWST)| flow cytometry/CD3"
cat(sprintf("\n\n### %s\n\n", mmain))
if(save_plots) png(file = "prcomp_CD3.png", width = width_png, height = width_png)
plot(pprcomp$x, col = as.character(annotation.df$CD3.col), main = mmain, 
     xlab = "first principal component", ylab = "secondo principal component", pch = 19)
ffill2 <- names(table(annotation.df$CD3.col))
color.bar2(-2.3, -0.7, ffill2, -3, values = vvalues)
text(x_text, y_text, "(a)")

mmain <-  "principal component analysis - CD19"#  "prcomp (AWST)| flow cytometry/CD19"
cat(sprintf("\n\n### %s\n\n", mmain))
if(save_plots) png(file = "prcomp_CD19.png", width = width_png, height = width_png)
plot(pprcomp$x, col = as.character(annotation.df$CD19.col), main = mmain, 
     xlab = "first principal component", ylab = "secondo principal component", pch = 19)
color.bar2(-2.3, -0.7, ffill2, -3, values = vvalues)
text(x_text, y_text, "(b)")

mmain <-  "principal component analysis - CD11c"#  "prcomp (AWST)| flow cytometry/CD11c"
cat(sprintf("\n\n### %s\n\n", mmain))
if(save_plots) png(file = "prcomp_CD11c.png", width = width_png, height = width_png)
plot(pprcomp$x, col = as.character(annotation.df$CD11c.col), main = mmain, 
     xlab = "first principal component", ylab = "secondo principal component", pch = 19)
color.bar2(-2.3, -0.7, ffill2, -3, values = vvalues)
text(x_text, y_text, "(c)")


mmain <-  "principal component analysis - CD14"#  "prcomp (AWST)| flow cytometry/CD14"
cat(sprintf("\n\n### %s\n\n", mmain))
if(save_plots) png(file = "prcomp_CD14.png", width = width_png, height = width_png)
plot(pprcomp$x, col = as.character(annotation.df$CD14.col), main = mmain, 
     xlab = "first principal component", ylab = "secondo principal component", pch = 19)
color.bar2(-2.3, -0.7, ffill2, -3, values = vvalues)
text(x_text, y_text, "(d)")

mmain <-  "principal component analysis - CD4"#  "prcomp (AWST)| flow cytometry/CD4"
cat(sprintf("\n\n### %s\n\n", mmain))
if(save_plots) png(file = "prcomp_CD4.png", width = width_png, height = width_png)
plot(pprcomp$x, col = as.character(annotation.df$CD4.col), main = mmain, 
     xlab = "first principal component", ylab = "secondo principal component", pch = 19)
color.bar2(-2.3, -0.7, ffill2, -3, values = vvalues)
text(x_text, y_text, "(e)")

mmain <-  "principal component analysis - CD8"#  "prcomp (AWST)| flow cytometry/CD8"
cat(sprintf("\n\n### %s\n\n", mmain))
if(save_plots) png(file = "prcomp_CD8.png", width = width_png, height = width_png)
plot(pprcomp$x, col = as.character(annotation.df$CD8.col), main = mmain, 
     xlab = "first principal component", ylab = "secondo principal component", pch = 19)
color.bar2(-2.3, -0.7, ffill2, -3, values = vvalues)
text(x_text, y_text, "(f)")
```

# Rtsne p100 (AWST) | ADT

```{r Rtsne p100 (AWST)| cibersort, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center"}
load("expression_Rtsne_2d.RData")
ans_Rtsne$Y <- scale(ans_Rtsne$Y)
if(save_plots) png("Rtsne.png", width= png_width_small, height= png_height_small, res = png_res)
mmain = "T-distributed Stochastic Neighbor Embedding (tsne)"
plot(-ans_Rtsne$Y[, 1], ans_Rtsne$Y[, 2], col = clustering.df$CBMC6.col, main = mmain, xlab = "", ylab = "")
legend(-1.65, -1.5, legend=llegend, fill = ffill, y.intersp = 1, box.col = "white", border = "white", title = "CBMC", title.adj = 0)
text(-1.65, 2, "(c)")
```

# umap (AWST) | ADT
```{r umap 2d (AWST)| cibersort, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center"}
load("expression_umap_2d.RData")
ans_umap$layout <- scale(ans_umap$layout)
if(save_plots) png("umap.png", width= png_width_small, height= png_height_small, res = png_res)
mmain = "Uniform Manifold Approximation and Projection (umap)"
plot(-ans_umap$layout[, 2], -ans_umap$layout[, 1], col = clustering.df$CBMC6.col, main = mmain, xlab = "", ylab = "")
legend(-1.65, -1.5, legend=llegend, fill = ffill, y.intersp = 1, box.col = "white", border = "white", title = "CBMC", title.adj = 0)
text(-1.65, 1.5, "(d)")
```

# Session info
```{r}
sessionInfo()
```

# References

- [Simultaneous epitope and transcriptome measurement in single cells](https://www.nature.com/articles/nmeth.4380)
