---
title: "Code to reproduce the analyses of the AWST paper: Analysis of the SEQC data"
author: "[S.M.Pagnotta](https://www.researchgate.net/profile/Stefano_Pagnotta) - Dept. of Science and Technology, Universita' degli Studi del Sannio (Benevento, Italy)</br> [D.Risso](https://www.researchgate.net/profile/Davide_Risso) - Dept. of Statistical Sciences, Universita' degli Studi di Padova (Padova, Italy)"
date:  '`r format(Sys.time(), "%d %B, %Y")`'
output: 
  html_document: 
    toc: yes
    toc_float: yes
    df_print: paged
fontsize: 11pt
editor_options: 
  chunk_output_type: console
---

Here we show the code to reproduce the analyses of: Risso and Pagnotta (2020). Per-sample standardization and asymmetric winsorization lead to accurate classification of RNA-seq expression profiles. In preparation.

This file belongs to the repository: https://github.com/drisso/awst_analysis.

The code is released with license GPL v3.0.

```{r knitr, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      fig.height = 4.5, 
                      fig.width = 11, 
                      fig.align = "center")
```

# Install and load `awst`

```{r install_awst, eval=FALSE, echo=TRUE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("drisso/awst")
```

# Data import and cleaning

The collection of the SEQC datasets is available throught the `seqc` Bioconductor package.
It can be installed with the following.

```{r install_seqc, eval=FALSE, echo=TRUE}
BiocManager::install("seqc")
```

```{r setup, echo=FALSE, message=FALSE}
library(knitr)
library(cluster)
library(dendextend)
library(awst)
library(ConsensusClusterPlus)
#################################
png_width <- 600
png_height <- 300
png_res <- 1/300
save_png <- FALSE
x_legend <- 81
```

We next build the data matrix from the "ILM_aceview" experiments. We remove duplicate gene symbols, ERCC spike-ins, and genes with no ENTREZ ID.

```{r preproc, eval=TRUE}
if(!file.exists("seqc20190914_ILM_aceview_Level_3.RData")) {
  library(seqc)
  
  sites <- c("MAY", "BGI", "CNL", "NVS", "AGR", "COH") 
  k <- 1
  theSite <- sites[k]
  
  ddata <- get(paste0("ILM_aceview_gene_", theSite))
  ddata <- ddata[!is.na(ddata$EntrezID),]
  ddata <- ddata[!ddata$IsERCC,]
  feature_annotation <- data.frame(ddata[, 1:3], row.names = ddata$EntrezID)
  ddata <- as.matrix(ddata[, -c(1:4)])
  sum(duplicated(feature_annotation$Symbol))
  sum(duplicated(feature_annotation$EntrezID))
  row.names(ddata) <- feature_annotation$EntrezID
  
  wwhich <- which(substr(colnames(ddata), 1, 1) %in% c("E", "F"))
  dim(ddata <- ddata[, -wwhich])
  substr_colnames_ddata <- substr(colnames(ddata), 1, 3)
  table(substr_colnames_ddata)
  samples <- unique(substr_colnames_ddata)
  wwhich <- which(substr_colnames_ddata == samples[1])
  data_matrix <- matrix(rowMeans(ddata[, wwhich]), ncol = 1)
  rownames(data_matrix) <- rownames(ddata)
  for(kk in 2:length(samples)) {
    wwhich <- which(substr_colnames_ddata == samples[kk])
    data_matrix <- cbind(data_matrix, rowMeans(ddata[, wwhich]))
  }
  colnames(data_matrix) <- paste(gsub("_", "", samples), theSite, sep = "_")
  
  for(k in 2:length(sites)) {
    (theSite <- sites[k])
    ddata <- get(paste0("ILM_aceview_gene_", theSite))
    ddata <- ddata[!is.na(ddata$EntrezID),]
    ddata <- ddata[!ddata$IsERCC,]
    row.names(ddata) <- ddata$EntrezID
    ddata <- as.matrix(ddata[, -c(1:4)])
    ddata <- ddata[rownames(data_matrix),]
    
    wwhich <- which(substr(colnames(ddata), 1, 1) %in% c("E", "F"))
    if(length(wwhich) > 0) dim(ddata <- ddata[, -wwhich])
    substr_colnames_ddata <- substr(colnames(ddata), 1, 3)
    table(substr_colnames_ddata)
    samples <- unique(substr_colnames_ddata)
    table(substr_colnames_ddata)
    
    for(kk in 1:length(samples)) {
      wwhich <- which(substr_colnames_ddata == samples[kk])
      data_matrix <- cbind(data_matrix, rowMeans(ddata[, wwhich]))
    }
    which(colnames(data_matrix) == "")
    colnames(data_matrix)[which(colnames(data_matrix) == "")] <- paste(gsub("_", "", samples), theSite, sep = "_")
  }
  ddata <- floor(data_matrix)
  
  annotation.df <- data.frame(samples = colnames(ddata), row.names = colnames(ddata))
  annotation.df$sample <- substr(annotation.df$samples, 1, 1)
  annotation.df$sample.col <- factor(annotation.df$sample)
  levels(annotation.df$sample.col) <- clust.col <- c("black", "red", "green2", "blue")
  annotation.df$replicate <- substr(annotation.df$samples, 2, 2)
  annotation.df$site <- substr(annotation.df$samples, 4, 6)
  annotation.df$site.col <- factor(annotation.df$site)
  levels(annotation.df$site.col) <- palette()[2:7]
  kable(addmargins(table(annotation.df$sample, annotation.df$site)))
  
  save(ddata, annotation.df, feature_annotation, clust.col, 
       file = "seqc20190914_ILM_aceview_Level_3.RData")
} else {
  load("seqc20190914_ILM_aceview_Level_3.RData")
}

kable(addmargins(table(annotation.df$sample, annotation.df$site)), caption = "Distribution of samples per sites")
```

```{r additional functions, echo = FALSE}
source("functions.R")
```

# Figure 2 and Supplementary figure 1

## Figure 2a and Supplementary figure 1a: clustering on RSEM data after awst

```{r _a.png}
exprData <- awst(ddata)
mmain <- paste0(nrow(exprData), " samples/", ncol(exprData), " genes")
hhc <- hclust(ddist <- dist(exprData), method = "ward.D2")
aCalinski <- calinski(hhc, gMax = 6)
hhc$height <- hhc$height/max(hhc$height)
if(save_png) png(file = "figure2_a.png", width = png_width, height = png_height, res = png_res)
plot(hhc, hang = -1, labels = FALSE, xlab = "", sub = "", main = mmain)
###
wwhere <- 4
hh <- mean(c(hhc$height[length(hhc$height)-wwhere+2], hhc$height[length(hhc$height)-wwhere+1]))
abline(h = hh, col = "red")
clustering <- as.factor(cutree(hhc, k = wwhere))
names(clustering) <- hhc$labels

levels(clustering) <- clust.col
tmp.col <- cbind(as.character(clustering), as.character(annotation.df$sample.col), as.character(annotation.df$site.col))
colnames(tmp.col) <- c("clust", "sample", "site")
colored_bars(colors = tmp.col, dend = as.dendrogram(hhc), y_scale = -0.15, y_shift = -0.2)
llegend  <- c("A (UHRR)", "B (HBRR)", "C (0.75A+0.25B)", "D (0.25A+0.75B)")
legend(x_legend, 1, legend = llegend, fill = levels(annotation.df$sample.col), y.intersp = 1, box.col = "white", border = "white", title = "samples", title.adj = 0)
text(1, 0.95, "(a)")
add_Calinski_curve()
if(save_png) png(file = "suppFigure1_a.png", width = png_width, height = png_height)
ssil <- silhouette(as.numeric(clustering), ddist)
plot(ssil, main = mmain, col = clust.col)
```

## Figure 2b and Supplementary figure 1b: clustering on TPM data after Hart transformation

```{r _h.png}
gene_length <- feature_annotation[rownames(ddata), "GeneLength"]
names(gene_length) <- feature_annotation$EntrezID
exprData <- get_TPM(ddata, gene_length)
exprData <- get_Hart2013(exprData)
mmain <- paste0(nrow(exprData), " samples/", ncol(exprData), " genes")
###
hhc <- hclust(ddist <- dist(exprData), method = "ward.D2")
aCalinski <- calinski(hhc, gMax = 6)
hhc$height <- hhc$height/max(hhc$height)
if(save_png) png(file = "figure2_b.png", width = png_width, height = png_height, res = png_res)
plot(hhc, hang = -1, labels = FALSE, xlab = "", sub = "", main = mmain)
###clustering#############################################################################
wwhere <- 4
hh <- mean(c(hhc$height[length(hhc$height)-wwhere+2], hhc$height[length(hhc$height)-wwhere+1]))
abline(h = hh, col = "red")
clustering <- as.factor(cutree(hhc, k = wwhere))
names(clustering) <- hhc$labels
levels(clustering) <- clust.col
tmp.col <- cbind(as.character(clustering), as.character(annotation.df$sample.col), as.character(annotation.df$site.col))
colnames(tmp.col) <- c("clust", "sample", "site")
colored_bars(colors = tmp.col, dend = as.dendrogram(hhc), y_scale = -0.15, y_shift = -0.2)
legend(x_legend, 1, legend=llegend, fill = levels(annotation.df$sample.col), y.intersp = 1, box.col = "white", border = "white", title = "samples", title.adj = 0)
text(1, 0.95, "(h)")
add_Calinski_curve()
if(save_png) png(file = "suppFigure1_b", width = png_width, height = png_height)
ssil <- silhouette(as.numeric(clustering), ddist)
plot(ssil, main = mmain, col = clust.col)
```

## Figure 2c: clustering on CPM data (std values)

```{r _c.png}
exprData <-  get_CPM(ddata)
exprData <-  scale(exprData)
mmain <- paste0(nrow(exprData), " samples/", ncol(exprData), " genes")
###
hhc <- hclust(ddist <- dist(exprData), method = "ward.D2")
aCalinski <- calinski(hhc, gMax = 6)
attr(aCalinski, "message") <- ""
hhc$height <- hhc$height/max(hhc$height)
if(save_png) png(file = "Figure2_c.png", width = png_width, height = png_height, res = png_res)
plot(hhc, hang = -1, labels = FALSE, xlab = "", sub = "", main = mmain)
wwhere <- 2
hh <- mean(c(hhc$height[length(hhc$height)-wwhere+2], hhc$height[length(hhc$height)-wwhere+1]))
abline(h = hh, col = "red")
clustering <- as.factor(cutree(hhc, k = wwhere))
names(clustering) <- hhc$labels
levels(clustering) <- clust.col
tmp.col <- cbind(as.character(clustering), as.character(annotation.df$sample.col), as.character(annotation.df$site.col))
colnames(tmp.col) <- c("clust", "sample", "site")
colored_bars(colors = tmp.col, dend = as.dendrogram(hhc), y_scale = -0.15, y_shift = -0.2)
legend(x_legend, 1, legend=llegend, fill = levels(annotation.df$sample.col), y.intersp = 1, box.col = "white", border = "white", title = "samples", title.adj = 0)
text(1, 0.95, "(c)")
add_Calinski_curve()
if(save_png) png(file = "suppFigure1_c.png", width = png_width, height = png_height, res = png_res)
ssil <- silhouette(as.numeric(clustering), ddist)
plot(ssil, main = paste0("(c) ", mmain), col = c("black", "red"))
```

## Figure 2d and Supplementary figure 1d: clustering on CPM data (std values; top 1000 genes)

```{r _d.png}
exprData <-  get_CPM(ddata)
tmp <- apply(exprData, 2, sd)
tmp <- sort(tmp)
tmp <- names(tail(tmp, 1000))
exprData <-  scale(exprData[, tmp])
mmain <- paste0(nrow(exprData), " samples/", ncol(exprData), " genes")
###
hhc <- hclust(ddist <- d_3c <- dist(exprData), method = "ward.D2")
aCalinski <- calinski(hhc, gMax = 6)
hhc$height <- hhc$height/max(hhc$height)
if(save_png) png(file = "Figure2_d.png", width = png_width, height = png_height, res = png_res)
plot(hhc, hang = -1, labels = FALSE, xlab = "", sub = "", main = mmain)
###clustering#############################################################################
wwhere <- 4
hh <- mean(c(hhc$height[length(hhc$height)-wwhere+2], hhc$height[length(hhc$height)-wwhere+1]))
abline(h = hh, col = "red")
clustering <- as.factor(cutree(hhc, k = wwhere))
names(clustering) <- hhc$labels
levels(clustering) <- clust.col
tmp.col <- cbind(as.character(clustering), as.character(annotation.df$sample.col), as.character(annotation.df$site.col))
colnames(tmp.col) <- c("clust", "sample", "site")
colored_bars(colors = tmp.col, dend = as.dendrogram(hhc), y_scale = -0.15, y_shift = -0.2)
legend(x_legend, 1, legend=llegend, fill = levels(annotation.df$sample.col), y.intersp = 1, box.col = "white", border = "white", title = "samples", title.adj = 0)
text(1, 0.95, "(d)")
add_Calinski_curve()
if(save_png) png(file = "suppFigure1_d.png", width = png_width, height = png_height)
ssil <- silhouette(as.numeric(clustering), ddist)
plot(ssil, main = mmain, col = clust.col)
```

# Supplementary Figures 2 and 3

```{r setup_of_figure_3}
noof_topVariableGenes <- 1000
set.seed(20190507) # needed to edit the color in the figures
#set.seed(Sys.time())
head(randomizedSamples <- sample(as.character(annotation.df$samples)))
annotation.df$noised <- "white"
noise_col <- "gray40"
mmark <- ""
ddf <- 4
noof_genes <- floor(nrow(ddata)/3)
apply_perturbation <- function() return(0.1 + rchisq(noof_genes, df = ddf))
```

## 3 perturbed samples

```{r perturb 3 sample}
# perturbing the original raw-counts
for(kk in 1:3) {
  someGenes <- sample(rownames(ddata), noof_genes, replace = FALSE)
  tmp <- apply_perturbation() * ddata[someGenes, randomizedSamples[kk]]
  ddata[someGenes, randomizedSamples[kk]] <- tmp
  annotation.df$noised[which(annotation.df$samples == randomizedSamples[kk])] <- noise_col
}
k <- kk + 1
```


### Supplementary Figures 2a and 3a: clustering on CPM data (std values; top 1000 genes)

```{r exprData_CPM}
exprData_CPM <- get_CPM(ddata)
apply_exprData_sd <- sort(apply(exprData_CPM, 2, sd))
topVariableGenes <- names(tail(apply_exprData_sd, noof_topVariableGenes))
exprData <-  scale(exprData_CPM[, topVariableGenes])
mmain <- paste0(nrow(exprData), " samples/", ncol(exprData), " genes")
hhc <- hclust(ddist <- dist(exprData), method = "ward.D2")
mmark <- "a"
hhc_3a <- as.dendrogram(hhc)
```

```{r cluster_CPM, results="hide"}
aCalinski <- calinski(hhc, gMax = 10)
hhc$height <- hhc$height/max(hhc$height)
if(save_png) png(file = paste0("suppFigure2_", mmark, ".png"), width = png_width, height = png_height, res = png_res)
plot(hhc, hang = -1, labels = FALSE, xlab = "", sub = "", main = mmain)
###clustering#############################################################################
wwhere <- 2
hh <- mean(c(hhc$height[length(hhc$height)-wwhere+2], hhc$height[length(hhc$height)-wwhere+1]))
#abline(h = hh, col = "red")
clustering <- as.factor(cutree(hhc, k = wwhere))
names(clustering) <- hhc$labels
#table(ssample.col, clustering[names(ssample.col)])
levels(clustering)[which(levels(clustering) == "1")] <- "black"
levels(clustering)[which(levels(clustering) == "2")] <- "red"
#levels(clustering)[which(levels(clustering) == "3")] <- "green3"
#levels(clustering)[which(levels(clustering) == "4")] <- "blue"
tmp.col <- cbind(annotation.df$noised, annotation.df$site.col, 
                 annotation.df$sample.col, clustering)
colnames(tmp.col) <- c("noise", "site", "sample", "clust")
colored_bars(colors = tmp.col, dend = as.dendrogram(hhc), y_scale = 0.2, y_shift = -0.01)
legend(x_legend, 1, legend=llegend, fill = clust.col, y.intersp = 1, box.col = "white", border = "white", title = "samples", title.adj = 0)
text(1, 0.95, paste0("(", mmark, ")"))
add_Calinski_curve()
ssil <- silhouette(as.numeric(annotation.df$sample.col), ddist)
if(save_png) png(file = paste0("suppFigure3_", mmark, ".png"), width = png_width, height = png_height, res = png_res)
plot(ssil, main = paste0("(", mmark, ") ", mmain), col = clust.col)
```

### Supplementary Figures 2e and 3e: clustering on RSEM data after awst

```{r RSEM_AWST, results="hide"}
exprData <-  awst(ddata)
mmain <- paste0(nrow(exprData), " samples/", ncol(exprData), " genes")
hhc <- hclust(ddist <- dist(exprData), method = "ward.D2")
pprcomp <- prcomp(exprData); pprcomp$x <- scale(pprcomp$x)
mmark <- "e"
hhc_3e <- as.dendrogram(hhc)
```

```{r plot_awst}
aCalinski <- calinski(hhc, gMax = 10)
hhc$height <- hhc$height/max(hhc$height)
if(save_png) png(file = paste0("suppFigure2_", mmark, ".png"), width = png_width, height = png_height, res = png_res)
plot(hhc, hang = -1, labels = FALSE, xlab = "", sub = "", main = mmain)
###clustering#############################################################################
wwhere <- 4
hh <- mean(c(hhc$height[length(hhc$height)-wwhere+2], hhc$height[length(hhc$height)-wwhere+1]))
clustering <- as.factor(cutree(hhc, k = wwhere))
names(clustering) <- hhc$labels
levels(clustering) <- clust.col
tmp.col <- cbind(annotation.df$noised, annotation.df$site.col, 
                 annotation.df$sample.col, clustering)
colnames(tmp.col) <- c("noise", "site", "sample", "clust")
colored_bars(colors = tmp.col, dend = as.dendrogram(hhc), y_scale = 0.2, y_shift = -0.01)
legend(x_legend, 1, legend = llegend, fill = levels(annotation.df$sample.col), y.intersp = 1, box.col = "white", border = "white", title = "samples", title.adj = 0)
text(1, 0.95, paste0("(", mmark, ")"))
add_Calinski_curve()
ssil <- silhouette(as.numeric(annotation.df$sample.col), ddist)
if(save_png) png(file = paste0("suppFigure3_", mmark, ".png"), width = png_width, height = png_height, res = png_res)
plot(ssil, main = paste0("(", mmark, ") ", mmain), col = clust.col)
```

## One-third perturbed samples

```{r perturb one-third samples}
# perturbing the original raw-counts
for(kk in k:36) {
  someGenes <- sample(rownames(ddata), noof_genes, replace = FALSE)
  tmp <- apply_perturbation() * ddata[someGenes, randomizedSamples[kk]]
  ddata[someGenes, randomizedSamples[kk]] <- tmp
  annotation.df$noised[which(annotation.df$samples == randomizedSamples[kk])] <- noise_col
}
k <- kk + 1
```

### Supplementary Figures 2b and 3b: clustering on CPM data (std values; top 1000 genes)

```{r ref.label="exprData_CPM"}
```
```{r, echo=FALSE}
hhc_3b <- as.dendrogram(hhc)
```
```{r}
mmark <- "b"
```
```{r ref.label="cluster_CPM"}
```


### Supplementary Figures 2f and 3f: clustering on RSEM data after awst

```{r, ref.label="RSEM_AWST", echo=FALSE, eval=TRUE}
```
```{r, echo=FALSE}
hhc_3f <- as.dendrogram(hhc)
```
```{r}
mmark <- "f"
```
```{r plot_ASWT_3f, echo=FALSE, eval=TRUE}
aCalinski <- calinski(hhc, gMax = 10)
hhc$height <- hhc$height/max(hhc$height)
if(save_png) png(file = paste0("suppFigure2_", mmark, ".png"), width = png_width, height = png_height, res = png_res)
plot(hhc, hang = -1, labels = FALSE, xlab = "", sub = "", main = mmain)
###clustering#############################################################################
wwhere <- 8
hh <- mean(c(hhc$height[length(hhc$height)-wwhere+2], hhc$height[length(hhc$height)-wwhere+1]))
#abline(h = hh, col = "red")
clustering <- as.factor(cutree(hhc, k = wwhere))
names(clustering) <- hhc$labels
#table(clustering, annotation.df$sample.col)
levels(clustering)[which(levels(clustering) == "1")] <- "black"
levels(clustering)[which(levels(clustering) == "2")] <- "gray"
levels(clustering)[which(levels(clustering) == "3")] <- "indianred"
levels(clustering)[which(levels(clustering) == "4")] <- "red"
levels(clustering)[which(levels(clustering) == "5")] <- "green3"
levels(clustering)[which(levels(clustering) == "6")] <- "blue"
levels(clustering)[which(levels(clustering) == "7")] <- "cornflowerblue"
levels(clustering)[which(levels(clustering) == "8")] <- "greenyellow"
clustering <- as.character(clustering)
tmp.col <- cbind(annotation.df$noised, annotation.df$site.col, 
                 annotation.df$sample.col, clustering)
colnames(tmp.col) <- c("noise", "site", "sample", "clust")
colored_bars(colors = tmp.col, dend = as.dendrogram(hhc), y_scale = 0.2, y_shift = -0.01)
legend(x_legend, 1, legend = llegend, fill = levels(annotation.df$sample.col), y.intersp = 1, box.col = "white", border = "white", title = "samples", title.adj = 0)
text(1, 0.95, paste0("(", mmark, ")"))
add_Calinski_curve()
ssil <- silhouette(as.numeric(annotation.df$sample.col), ddist)
if(save_png) png(file = paste0("suppFigure3_", mmark, ".png"), width = png_width, height = png_height, res = png_res)
plot(ssil, main = paste0("(", mmark, ") ", mmain), col = clust.col)
```

## Two-third perturbed samples

```{r perturb two-third samples}
# perturbing the original raw-counts
for(kk in k:72) {  
  someGenes <- sample(rownames(ddata), noof_genes, replace = FALSE)
  tmp <- apply_perturbation() * ddata[someGenes, randomizedSamples[kk]]
  ddata[someGenes, randomizedSamples[kk]] <- tmp
  annotation.df$noised[which(annotation.df$samples == randomizedSamples[kk])] <- noise_col
}
k <- kk + 1
```

### Supplementary Figures 2c and 3c: clustering on CPM data (std values; top 1000 genes)

```{r ref.label="exprData_CPM"}
```
```{r, echo=FALSE}
hhc_3c <- as.dendrogram(hhc)
```
```{r}
mmark <- "c"
```
```{r ref.label="cluster_CPM"}
```


### Supplementary Figures 2g and 3g: clustering on RSEM data after awst

```{r, ref.label="RSEM_AWST", echo=FALSE, eval=TRUE}
```
```{r, echo=FALSE}
hhc_3g <- as.dendrogram(hhc)
```
```{r}
mmark <- "g"
```
```{r plot_ASWT_3g, echo=FALSE, eval=TRUE}
aCalinski <- calinski(hhc, gMax = 10)
hhc$height <- hhc$height/max(hhc$height)
if(save_png) png(file = paste0("suppFigure2_", mmark, ".png"), width = png_width, height = png_height, res = png_res)
plot(hhc, hang = -1, labels = FALSE, xlab = "", sub = "", main = mmain)
###clustering#############################################################################
wwhere <- 8
hh <- mean(c(hhc$height[length(hhc$height)-wwhere+2], hhc$height[length(hhc$height)-wwhere+1]))
#abline(h = hh, col = "red")
clustering <- as.factor(cutree(hhc, k = wwhere))
names(clustering) <- hhc$labels
#table(clustering, annotation.df$sample.col)
levels(clustering)[which(levels(clustering) == "1")] <- "black"
levels(clustering)[which(levels(clustering) == "2")] <- "gray"
levels(clustering)[which(levels(clustering) == "3")] <- "indianred"
levels(clustering)[which(levels(clustering) == "4")] <- "red"
levels(clustering)[which(levels(clustering) == "5")] <- "green3"
levels(clustering)[which(levels(clustering) == "6")] <- "greenyellow"
levels(clustering)[which(levels(clustering) == "7")] <- "blue"
levels(clustering)[which(levels(clustering) == "8")] <- "cornflowerblue"
clustering <- as.character(clustering)
tmp.col <- cbind(annotation.df$noised, annotation.df$site.col, 
                 annotation.df$sample.col, clustering)
colnames(tmp.col) <- c("noise", "site", "sample", "clust")
colored_bars(colors = tmp.col, dend = as.dendrogram(hhc), y_scale = 0.2, y_shift = -0.01)
legend(x_legend, 1, legend = llegend, fill = levels(annotation.df$sample.col), y.intersp = 1, box.col = "white", border = "white", title = "samples", title.adj = 0)
text(1, 0.95, paste0("(", mmark, ")"))
add_Calinski_curve()
ssil <- silhouette(as.numeric(annotation.df$sample.col), ddist)
if(save_png) png(file = paste0("suppFigure3_", mmark, ".png"), width = png_width, height = png_height, res = png_res)
plot(ssil, main = paste0("(", mmark, ") ", mmain), col = clust.col)
```

## All perturbed samples

```{r perturb all samples}
# perturbing the original raw-counts
for(kk in k:108) { 
  someGenes <- sample(rownames(ddata), noof_genes, replace = FALSE)
  tmp <- apply_perturbation() * ddata[someGenes, randomizedSamples[kk]]
  ddata[someGenes, randomizedSamples[kk]] <- tmp
  annotation.df$noised[which(annotation.df$samples == randomizedSamples[kk])] <- noise_col
}
```

### Supplementary Figures 2d and 3d: clustering on CPM data (std values; top 1000 genes)

```{r ref.label="exprData_CPM"}
```
```{r, echo=FALSE}
hhc_3d <- as.dendrogram(hhc)
```
```{r}
mmark <- "d"
```
```{r ref.label="cluster_CPM"}
```

### Supplementary Figures 2h and 3h: clustering on RSEM data after awst

```{r, ref.label="RSEM_AWST", echo=FALSE, eval=TRUE}
```
```{r, echo=FALSE}
hhc_3h <- as.dendrogram(hhc)
```
```{r}
mmark <- "h"
```
```{r, ref.label="plot_awst"}
```

# Figure 3

## Radovich procedure

```{r, echo=FALSE}
#After restricting to genes with at least 75% non-zero RSEM values,
tmp <- rowMeans(ddata > 0, na.rm = TRUE)
dim(rdata <- ddata[which(tmp >= 0.75),])
#he genes with the 1000 highest median absolute deviation (MAD) values were chosen.
tmp <- tail(sort(apply(rdata, 1, mad)), 1000)
dim(rdata <- rdata[names(tmp),])
#RSEM values identically equal to zero were replaced into smallest non-zero value.
rdata[which(rdata == 0)] <- min(rdata[rdata > 0])
#Then a log2 transformation was applied
rdata <- log2(rdata)
#and the values were median centered by gene and divided by MAD expression of each gene.
mmedian <- apply(rdata, 1, median)
mmad <- apply(rdata, 1, mad)
rdata <- t(scale(t(rdata), center = mmedian, scale = mmad))
```

```{r, echo=FALSE, eval = TRUE}
set.seed(20200127)
library(cluster)
library(ConsensusClusterPlus)
if(save_png) ans <- ConsensusClusterPlus(rdata, maxK = 5, reps = 1000, distance = "euclidean", clusterAlg = "pam", title = "Radovich2018", plot = "png") else ans <- ConsensusClusterPlus(rdata, maxK = 5, reps = 1000, distance = "euclidean", clusterAlg = "pam")
rownames(annotation.df) <- annotation.df$samples
annotation.df <- annotation.df[names(ans[[2]]$consensusClass),]
vCramer(ans[[2]]$consensusClass, annotation.df$sample)
if("consensus" %in% .packages(TRUE)) {
  if(save_png) png(file = "Radovich2018_cpm_2.png", width = png_width, height = png_width, res = png_res)
  class(ans[[2]]) <- "ConsensusClusterPlus"
  plot(ans[[2]], G = 2, rep = 1000, clustering = annotation.df$sample.col, original = TRUE)
}

annotation.df <- annotation.df[names(ans[[4]]$consensusClass),]
vCramer(ans[[4]]$consensusClass, annotation.df$sample)
if("consensus" %in% .packages(TRUE)) {
  if(save_png) png(file = "Radovich2018_cpm_4.png", width = png_width, height = png_width, res = png_res)
  class(ans[[4]]) <- "ConsensusClusterPlus"
  plot(ans[[4]], G = 4, rep = 1000, clustering = annotation.df$sample.col, original = TRUE)
}
```

## Consensus with AWST

```{r, echo=FALSE}
exprData <- awst(ddata)
if(save_png) ans <- ConsensusClusterPlus(t(exprData), maxK = 5, reps = 1000, distance = "euclidean", clusterAlg = "pam", title = "Radovich2018_awst", plot = "png") else ans <- ConsensusClusterPlus(t(exprData), maxK = 5, reps = 1000, distance = "euclidean", clusterAlg = "pam")
rownames(annotation.df) <- annotation.df$samples
annotation.df <- annotation.df[names(ans[[2]]$consensusClass),]
vCramer(ans[[2]]$consensusClass, annotation.df$sample)
if("consensus" %in% .packages(TRUE)) {
  if(save_png) png(file = "Radovich2018_awst_2.png", width = png_width, height = png_width, res = png_res)
  class(ans[[2]]) <- "ConsensusClusterPlus"
  plot(ans[[2]], G = 2, rep = 1000, clustering = annotation.df$sample.col, original = TRUE)
}
annotation.df <- annotation.df[names(ans[[4]]$consensusClass),]
vCramer(ans[[4]]$consensusClass, annotation.df$sample)
if("consensus" %in% .packages(TRUE)) {
  if(save_png) png(file = "Radovich2018_awst_4.png", width = png_width, height = png_width, res = png_res)
  class(ans[[4]]) <- "ConsensusClusterPlus"
  plot(ans[[4]], G = 4, rep = 1000, clustering = annotation.df$sample.col, original = TRUE)
}
```

# Session info

```{r}
sessionInfo()
```

# References

- [A comprehensive assessment of RNA-seq accuracy, reproducibility and information content by the Sequencing Quality Control Consortium](https://www.nature.com/articles/nbt.2957)

- [RNA-seq data generated from SEQC (MAQC-III) study](http://bioconductor.org/packages/release/data/experiment/html/seqc.html)