---
title: "Code to reproduce the analyses of the AWST paper: MIXOLOGY"
author: "[S.M.Pagnotta](https://www.researchgate.net/profile/Stefano_Pagnotta) - Dept. of Science and Technology, Universita' degli Studi del Sannio (Benevento, Italy)</br> [D.Risso](https://www.researchgate.net/profile/Davide_Risso) - Dept. of Statistical Sciences, Universita' degli Studi di Padova (Padova, Italy)"
date:  '`r format(Sys.time(), "%d %B, %Y")`'
output: 
  html_document: 
    toc: yes
    toc_float: yes
    df_print: paged
fontsize: 11pt
editor_options: 
  chunk_output_type: console
---

```{r knitr, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE, fig.height = 4.5, fig.width =  11, fig.align = "center")
```

blah blah 

This file belongs to the repository: https://github.com/drisso/awst_analysis.

The code is released with license GPL v3.0.



# Benchmarking single cell RNA-sequencing analysis pipelines using mixture control experiments [(Nature)](https://www.nature.com/articles/s41592-019-0425-8)
### Supplementary Figure 3 Visualization of representative benchmarking datasets using t-SNE and UMAP and violin plot of the number of doublets.
 (a) t-SNE and UMAP visualizations of 4 datasets. From left to right: 10X single cell using 3 cell lines; 10X single cell using 5 cell lines; a CEL-seq2 cell mixture and a CEL-seq2 RNA mixture. Each point represents a cell or ‘pseudo cell’ and the number of cells in each plot is indicated in the title. (b) Violin plot of doublets in each dataset, identified using Demuxlet (DBL: doublet, SNG: single cell). The number of single cells and doublets are shown on top of each violin. Doublets were excluded when calculating the performance metrics.
![](https://media.springernature.com/full/springer-static/esm/art%3A10.1038%2Fs41592-019-0425-8/MediaObjects/41592_2019_425_Fig9_ESM.jpg?as=webp)


### Supplementary Figure 5: Example clustering results and summary of clustering performance using ARI and the number of clusters.
 (a) Examples of clustering results visualized by PCA (top) and t-SNE (bottom), with different colours representing the cluster assignments made by the selected method. The sample sizes are 340 for RNAmix_CEL-seq2, 285 for cellmix3 and 274 for sc_CEL-seq2. Coefficients obtained from linear models fitted using the ARI (b) or the number of clusters (c) as dependent variables, and experimental design, normalization methods, imputation methods and clustering methods as covariates. The coefficients measure whether particular features have positive or negative associations with the dependent variables.
![](https://media.springernature.com/full/springer-static/esm/art%3A10.1038%2Fs41592-019-0425-8/MediaObjects/41592_2019_425_Fig11_ESM.jpg?as=webp)

## The data are from [(github) single cell mixology: single cell RNA-seq benchmarking](https://github.com/LuyiTian/sc_mixology)
- [scRNA-seq mixology: towards better benchmarking of single cell RNA-seq analysis methods](https://www.biorxiv.org/content/10.1101/433102v3)



# Install and load `awst`

```{r install_awst, eval=FALSE, echo=TRUE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("drisso/awst")
#BiocManager::install("GIS-SP-Group/RCA")
#BiocManager::install("hgu133plus2.db")
```

# Setup
```{r setup, message=FALSE}
rm(list = ls())
#library(steFunctions)
library(dendextend)
library(clues)
#setwd("~/Dropbox/AWST/mixology/")
jobName <- "mixoloy20200818"
source(url("https://raw.githubusercontent.com/drisso/awst_analysis/master/functions.R"))
#####
save_plots <- FALSE
png_width_large <- 2100
png_height_large <- 500
png_width_small <- width_png <- 700
png_height_small <- 700
png_res <- 1/300
####
true_number_of_clusters <- c(3, 5, 7, 10, 10)
names(true_number_of_clusters) <- c("sc_10x", "sc_10x_5cl", "RNAmix_celseq2", "cellmix4", "cellmix3")
####
results <- matrix(NA, ncol = 8, nrow = 12)
colnames(results) <- c("where", "what", "ARI", "accuracy", "purity", "avg", "noOfClust_Est", "noOfClust_Th") 
#results <- data.frame(where = rep("", 30), what = "", ARI = NA, accuracy = NA, purity = NA, noOfClust = NA)
k <- 0
```

```{r additional functions, echo=FALSE}
cal_entropy=function(x){
  freqs <- table(x)/length(x)
  freqs = freqs[freqs>0]
  return(-sum(freqs * log(freqs))/log(length(x)))
}

get_cluster_purity=function(theretical, estimated){
  return(mean(unlist(lapply(unique(theretical), function(x){cal_entropy(estimated[theretical == x])}))))
}

get_cluster_accuracy=function(theretical, estimated){
  return(mean(unlist(lapply(unique(estimated), function(x){cal_entropy(theretical[estimated == x])})), na.rm = TRUE))
}

get_statistics <- function() {
  message("estimated no. of clusters: ", tmp <- length(unique(clustering.df$tmp)))
  tmp <- paste(tmp); if(min(table(clustering.df$tmp)) == 1) tmp <- paste0(tmp, "(*)")

  ARI <- adjustedRand(as.numeric(factor(clustering.df$tmp)),
                      as.numeric(factor(clustering.df$True.col)))[1]
  message("adjusted Rand-index: ", round(ARI, 4))

  cluster_purity <- get_cluster_purity(clustering.df$True.col, clustering.df$tmp)
  message("cluster purity: ", round(cluster_purity, 4))

  cluster_accuracy <- get_cluster_accuracy(clustering.df$True.col, clustering.df$tmp)
  message("cluster accuracy: ", round(cluster_accuracy, 4))
  avg <- round(exp(log(ARI * (1 - cluster_accuracy) * (1 - cluster_purity))/3), 4)
  ans <- c(prefix, what, ARI, cluster_accuracy, cluster_purity, 
           avg, tmp, true_number_of_clusters[prefix])
  
  return(ans)
}

scran_high_var = function(sce,topn=1000){
  require(scran)
  if(max(logcounts(sce)>100)){
    logcounts(sce) = log2(logcounts(sce)+1-min(min(logcounts(sce)),0))
  }
  var.fit <- trendVar(sce, method="loess", use.spikes=FALSE)
  var.out <- decomposeVar(sce, var.fit)
  hvg.out <- var.out[order(var.out$bio, decreasing=TRUE)[1:topn], ]
  rowData(sce)$hi_var = FALSE
  rowData(sce)$hi_var[rownames(rowData(sce)) %in% rownames(hvg.out)] = TRUE
  return(sce)
}
```

```{r get_clusterExperiment, echo = FALSE}
get_clusterExperiment <- function(input = "counts") {
  require(clusterExperiment)
  
  if(input == "awst") {
    which_assay <- "awst"
  } else if(input == "scry") {
    which_assay <- "counts"
  } else {
    which_assay <- "logcounts"
  }
  
  sce <- SingleCellExperiment(
    assays = list(
      counts = as.matrix(ddata),
      logcounts = log2(as.matrix(ddata) + 1),
      awst = t(exprData)
    ), 
    colData = annotation.df
  )
  
  if(input == "awst") {
    reducedDim(sce, "PCA") <- pprcomp$x
  } else if(input == "scry") {
    reducedDim(sce, "PCA") <- irlba::prcomp_irlba(t(as.matrix(ddata)), n=5)$x
  } else {
    reducedDim(sce, "PCA") <- irlba::prcomp_irlba(t(log2(as.matrix(ddata)+1)), n=5)$x
  }
  
  #sce <- scran_high_var(sce)
#  sce <- sce[rowData(sce)$hi_var,]

  se = RSEC(sce, isCount = FALSE, whichAssay = which_assay, minSizes=5,
                reduceMethod="PCA", nReducedDims=5, ncores=4, random.seed=176201,
                dendroReduce="PCA", dendroNDims = 5, 
                mergeMethod = "adjP", mergeDEMethod = "limma", 
                mergeCutoff = 0.1, mergeLogFCcutoff = 1)
  
  plotDendrogram(se, colData = "cell.col", whichClusters = 1)
  
  colData(sce)$clustering_res = primaryCluster(se)

  return(sce)
}

prefix <- "sc_10x_5cl"

```

# sc_10x_5cl
```{r, eval = file.exists("sc_10x_expression_working.RData"), echo=FALSE}
load("sc_10x_5cl_expression_working.RData")
# annotation.df <- annotation.df[hhc$labels,]
```

```{r _rsec.RData, eval=!file.exists("sc_10x_5cl_rsec_working.RData"), echo=TRUE}
library(awst)
library(EDASeq)
library(Rtsne)
library(umap)
library(SingleCellExperiment)
library(clusterExperiment)
############################
load(paste0(prefix, "_counts.RData"))
load(paste0(prefix, "_expression.RData"))
#######################
# no_of_detected_gene_per_sample <- colSums(ddata > 0) 
# fivenum(no_of_detected_gene_per_sample)
# ddata <- EDASeq::betweenLaneNormalization(as.matrix(ddata), which = "full", round = FALSE)
# # apply the AWS-transformation 
# tmp <- rowSums(ddata)
# sum(tmp > 0)
# tmp <- colSums(ddata)
# sum(tmp > 0)
# 
# exprData <- awst(ddata, poscount = TRUE, full_quantile = TRUE)
# sum(is.na(rowSums(exprData)))
# sum(is.na(colSums(exprData)))

#save(exprData, prefix, file = paste0(prefix, "_expression.RData"))
#dim(exprData <- gene_filter(exprData))
#write.table(exprData, file =  paste0(prefix, "_expression.tsv"), sep = "\t")

nrow_exprData <- nrow(exprData)
ncol_exprData <- ncol(exprData)

tmp <- get_clusterExperiment()
wwhich <- which(colnames(colData(tmp)) %in% colnames(annotation.df))
ans_clusterExp_raw <- colData(tmp)[, -wwhich, drop=FALSE]

tmp <- get_clusterExperiment(input = "awst")
wwhich <- which(colnames(colData(tmp)) %in% colnames(annotation.df))
ans_clusterExp_awst <- colData(tmp)[, -wwhich, drop=FALSE]

load(paste0(prefix, "_VST.RData"))
tmp <- get_clusterExperiment()
wwhich <- which(colnames(colData(tmp)) %in% colnames(annotation.df))
ans_clusterExp_vst <- colData(tmp)[, -wwhich, drop=FALSE]

load(paste0(prefix, "_scry.RData"))
ddata <- scryData_pearson
tmp <- get_clusterExperiment(input = "scry")
wwhich <- which(colnames(colData(tmp)) %in% colnames(annotation.df))
ans_clusterExp_scry <- colData(tmp)[, -wwhich, drop=FALSE]

save(nrow_exprData, ncol_exprData, annotation.df,
     pprcomp, ans_Rtsne, ans_umap, prefix, 
     ans_clusterExp_awst, ans_clusterExp_raw,
     ans_clusterExp_vst, ans_clusterExp_scry,
     file = paste0(prefix, "_rsec_working.RData"))

```

```{r clExp, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
load(paste0(prefix, "_rsec_working.RData"))
clustering.df <- data.frame(cell_line = annotation.df$cell_line, 
                            True.col = annotation.df$cell.col,
                            row.names = rownames(annotation.df))
if("mRNA_amount.col" %in% colnames(annotation.df)) clustering.df$mRNA_amount.col <-  annotation.df$mRNA_amount.col

cat("\n\n### counts + clusterExperiment of ", prefix, "\n")

clustering.df <- clustering.df[rownames(ans_clusterExp_raw),]
clustering.df$ClustExp_raw <- clustering.df$tmp <- ans_clusterExp_raw$clustering_res
what <- "clusterExperiment +  counts"

results[k <- k+1,] <- get_statistics()
###
cat("\n\n### awst + clusterExperiment of ", prefix, "\n")

clustering.df <- clustering.df[rownames(ans_clusterExp_awst),]
clustering.df$ClustExp_awst <- clustering.df$tmp <- ans_clusterExp_awst$clustering_res
what <- "clusterExperiment +  AWST"

results[k <- k+1,] <- get_statistics()

###
cat("\n\n### vst + clusterExperiment of ", prefix, "\n")

clustering.df <- clustering.df[rownames(ans_clusterExp_vst),]
clustering.df$ClustExp_vst <- clustering.df$tmp <- ans_clusterExp_vst$clustering_res
what <- "clusterExperiment +  VST"

results[k <- k+1,] <- get_statistics()

###
cat("\n\n### scry + clusterExperiment of ", prefix, "\n")

clustering.df <- clustering.df[rownames(ans_clusterExp_scry),]
clustering.df$ClustExp_scry <- clustering.df$tmp <- ans_clusterExp_scry$clustering_res
what <- "clusterExperiment +  Townes"

results[k <- k+1,] <- get_statistics()
```

# sc_10x
```{r, eval = file.exists("sc_10x_expression_working.RData"), echo=FALSE}
load("sc_10x_expression_working.RData")
# annotation.df <- annotation.df[hhc$labels,]
```

```{r sc_10x, eval = !file.exists("sc_10x_expression_working.RData"), echo=FALSE}
con <- gzcon(url("https://github.com/LuyiTian/sc_mixology/blob/master/data/csv/sc_10x.metadata.csv.gz?raw=true"))
annotation.df <- read.csv(textConnection(readLines(con)))
annotation.df$cell.col <- factor(annotation.df$cell_line)
levels(annotation.df$cell.col) <- c("red", "blue", "green3")
annotation.df$cell.col <- as.character(annotation.df$cell.col)

con <- gzcon(url("https://github.com/LuyiTian/sc_mixology/blob/master/data/csv/sc_10x.count.csv.gz?raw=true"))
ddata <- read.csv(textConnection(readLines(con)))

both <- intersect(colnames(ddata), rownames(annotation.df))
ddata <- ddata[, both]
annotation.df <- annotation.df[both,]
prefix <- "sc_10x"
save(ddata, annotation.df, prefix, file = paste0(prefix, "_counts.RData"))
```

```{r, ref.label= "_working.RData", eval= !file.exists("sc_10x_rsec_working.RData"), echo=FALSE}
```
```{r, ref.label="clExp", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```


# RNAmix_celseq2
```{r, eval = file.exists("RNAmix_celseq2_expression_working.RData"), echo=FALSE}
load("RNAmix_celseq2_expression_working.RData")
# annotation.df <- annotation.df[hhc$labels,]
```

```{r RNAmix_celseq2, eval = !file.exists("RNAmix_celseq2_expression_working.RData"), echo=FALSE}
con <- gzcon(url("https://github.com/LuyiTian/sc_mixology/blob/master/data/csv/RNAmix_celseq2.metadata.csv.gz?raw=true"))
annotation.df <- read.csv(textConnection(readLines(con)))
annotation.df$cell_line <- factor(annotation.df$mix)
levels(annotation.df$cell_line)[1:2] <- "1-2"
levels(annotation.df$cell_line) <- paste("mix", levels(annotation.df$cell_line))

annotation.df$cell.col <- annotation.df$cell_line
table(annotation.df$cell.col)
levels(annotation.df$cell.col) <- c("gray25", "green4", "cyan4", "red4", "green", "cyan", "red")
annotation.df$cell.col <- as.character(annotation.df$cell.col)


annotation.df$mRNA_amount.col <- factor(annotation.df$mRNA_amount)
levels(annotation.df$mRNA_amount.col) <- c("green1", "green2", "green3", "green4")
annotation.df$mRNA_amount.col <- as.character(annotation.df$mRNA_amount.col)


con <- gzcon(url("https://github.com/LuyiTian/sc_mixology/blob/master/data/csv/RNAmix_celseq2.count.csv.gz?raw=true"))
ddata <- read.csv(textConnection(readLines(con)))

both <- intersect(colnames(ddata), rownames(annotation.df))
ddata <- ddata[, both]
annotation.df <- annotation.df[both,]
prefix <- "RNAmix_celseq2"
save(ddata, annotation.df, prefix, file = paste0(prefix, "_counts.RData"))
```
```{r, ref.label= "_working.RData", eval= !file.exists("RNAmix_celseq2_rsec_working.RData"), echo=FALSE}
```
```{r, ref.label="clExp", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```


# Table
```{r, echo=FALSE}
results <- as.data.frame(results)
results <- results[!is.na(results$ARI),]
results$ARI <- round(as.numeric(as.character(results$ARI)), 4)
results$accuracy <- round(as.numeric(as.character(results$accuracy)), 4)
results$purity <- round(as.numeric(as.character(results$purity)), 4)

knitr::kable(results)

write.table(results, file = paste0(jobName, "clExp_results.tsv"), sep = "\t")
```

(*) with at least a cluster of size 1.

# Session info
```{r}
sessionInfo()
```


```{r}
knitr::knit_exit()
```

```{r}
rm(list = ls())
setwd("~/Dropbox/AWST/mixology/")
ttable <- read.csv("mixoloy20200818_results.tsv", sep = "\t")
protocol <- ttable[grep("protocol", ttable$what),]
ttable <- ttable[-grep("protocol", ttable$what),]
awst <- ttable[grep("AWST", ttable$what),]
raw <- ttable[grep("counts", ttable$what),]
K <- nrow(raw)

tmp <- c(raw$avg, awst$avg)
names(tmp) <- c(rep("raw", K), rep("awst", K))
boxplot(tmp ~ names(tmp))
t.test(tmp ~ names(tmp))
wilcox.test(tmp ~ names(tmp))

RaceID <- grep("RaceID", raw$what)
Seurat <- grep("Seurat", raw$what)
SC3 <- grep("SC3", raw$what)
RCA <- grep("RCA", raw$what)

RNAmix_celseq2 <- grep("RNAmix_celseq2", raw$where)
cellmix3 <- grep("cellmix3", raw$where)
cellmix4 <- grep("cellmix4", raw$where)
sc_10x_5cl <- grep("sc_10x_5cl", raw$where)
sc_10x <- grep("sc_10x", raw$where)
sc_10x <- setdiff(sc_10x, sc_10x_5cl)
llty <- rep(1, K)
llty[RNAmix_celseq2] <- 4
llty[sc_10x_5cl] <- 3
llty[cellmix3] <- 2
llty[cellmix4] <- 5

ccol <- rep("black", K)
ccol[RaceID] <- "green2"
ccol[RCA] <- "magenta"
ccol[Seurat] <- "red"
ccol[SC3] <- "blue"
plot(c(-0.5, 1.5), c(1, 1), col = "white", ylim = c(.5, 1))
segments(0, raw$avg, 1, awst$avg, col = ccol, lty = llty)
points(rep(1, K), awst$avg, col = ccol, pch = 19)
points(rep(0, K), raw$avg, col = ccol, pch = 19)
points(rep(1.05, nrow(protocol)), protocol$avg)
text(rep(1.05, nrow(protocol)), protocol$avg, protocol$where, pos = 4)
```

