---
title: "Code to reproduce the analyses of the AWST paper: MIXOLOGY<br/> Figure 4,<br/> Supplementary Table 3."
author: "[S.M.Pagnotta](https://www.researchgate.net/profile/Stefano_Pagnotta) - Dept. of Science and Technology, Universita' degli Studi del Sannio (Benevento, Italy)</br> [D.Risso](https://www.researchgate.net/profile/Davide_Risso) - Dept. of Statistical Sciences, Universita' degli Studi di Padova (Padova, Italy)"
date:  '`r format(Sys.time(), "%d %B, %Y")`'
output: 
  html_document: 
    toc: yes
    toc_float: yes
    df_print: paged
fontsize: 11pt
editor_options: 
  chunk_output_type: console
---

```{r knitr, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE, fig.height = 4.5, fig.width =  11, fig.align = "center")
```

This file belongs to the repository: https://github.com/drisso/awst_analysis.

The code is released with license GPL v3.0.



# Benchmarking single cell RNA-sequencing analysis pipelines using mixture control experiments [(Nature)](https://www.nature.com/articles/s41592-019-0425-8)

## The data are from [(github) single cell mixology: single cell RNA-seq benchmarking](https://github.com/LuyiTian/sc_mixology)
- [scRNA-seq mixology: towards better benchmarking of single cell RNA-seq analysis methods](https://www.biorxiv.org/content/10.1101/433102v3)



# Install and load `awst`

```{r install_awst, eval=FALSE, echo=TRUE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("drisso/awst")
#BiocManager::install("GIS-SP-Group/RCA")
#BiocManager::install("hgu133plus2.db")
```

# Setup
```{r setup, message=FALSE}
rm(list = ls())
library(steFunctions)
library(dendextend)
library(clues)
setwd("~/Dropbox/AWST/mixology/")
jobName <- "mixoloy20200925"
source(url("https://raw.githubusercontent.com/drisso/awst_analysis/master/functions.R"))
#####
save_plots <- FALSE
png_width_large <- 2100
png_height_large <- 500
png_width_small <- width_png <- 700
png_height_small <- 700
png_res <- 1/300
####
results <- matrix(NA, ncol = 8, nrow = 170)
colnames(results) <- c("where", "what", "eca", "ecp", "ari", "G", "noOfClust_Th", "noOfClust_Est") 
k <- 0
```

```{r get_RaceID, echo=TRUE}
get_RaceID <- function(eData, is_normalized = FALSE) {
  require(RaceID)
  
  
  if(!is_normalized) eData <- log2(as.matrix(eData) + 1)
    
  sc <- SCseq(as.data.frame(eData))
  
  sc@ndata = sc@expdata
  sc@genes = rownames(sc@ndata)
  sc@counts = rep(1,ncol(sc@ndata))
  names(sc@counts) = colnames(sc@ndata)
  sc@cluster$features = sc@genes

  sc <- compdist(sc, metric="pearson", FSelect = FALSE, knn = NULL)
  sc <- clustexp(sc, sat = TRUE, samp = NULL, cln = NULL, clustnr = 30,
               bootnr = 50, rseed = 17000, FUNcluster = "kmedoids")

  annotation.df$clustering = as.factor(sc@cluster$kpart)
  
  sc <- compumap(sc)
  tmp <- scale(sc@umap)
  annotation.df$umap.1 <- tmp[, 1]
  annotation.df$umap.2 <- tmp[, 2]

  sc <- comptsne(sc)
  tmp <- scale(sc@tsne)
  annotation.df$tsne.1 <- tmp[, 1]
  annotation.df$tsne.2 <- tmp[, 2]

  return(annotation.df)
}
```
```{r get_RCA, echo = TRUE}
get_RCA <- function(eData, is_normalized = FALSE) {

  if(prefix == "sc_10x_5cl") {
    wwhich <- grep("ENSG00", rownames(eData))
    rownames(eData)[wwhich] <- gsub("ENSG0", "", rownames(ddata)[wwhich])
    rownames(eData) <- paste("XXXX",rownames(eData),sep="_")
  } else {
    library(biomaRt)
    mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
    G_list <- getBM(filters="ensembl_gene_id",
                    attributes=c("external_gene_name","ensembl_gene_id","external_gene_source"),
                    values=rownames(ddata), mart=mart)
    G_list = G_list[G_list$external_gene_source == "HGNC Symbol",]
    G_list$format_name =   paste("XXXX",G_list$external_gene_name,G_list$ensembl_gene_id,sep="_")
    rownames(G_list) = G_list$ensembl_gene_id

    both <- intersect(rownames(G_list), rownames(eData))
    eData <- eData[both,]
    G_list <- G_list[both,]

    rownames(eData) <- G_list$format_name
  }
  
  require(SingleCellExperiment)
  if(!is_normalized) {
    sce <- SingleCellExperiment(
      assays = list(
        counts = eData,
        logcounts = log2(eData + 1)
      ), colData = annotation.df) 
  } else {
    sce <- SingleCellExperiment(
      assays = list(
        counts = 2^eData,
        logcounts = eData
      ), colData = annotation.df)
  }
  
  require(preprocessCore)
  require(flashClust)
  require(RCA)
  
  data_obj = dataConstruct(as.matrix(logcounts(sce)))
  data_obj$geneFilter <- rep(TRUE, nrow(logcounts(sce)))
  data_obj$fpkm_transformed <- data_obj$fpkm_raw
  data_obj = featureConstruct(data_obj, method = "GlobalPanel")

  set.seed(20742579)
  data_obj = cellClust(data_obj, method = "hclust", deepSplit_wgcna = 1,
                       min_group_Size_wgcna = 5)

  annotation.df$clustering <- factor(data_obj$group_labels_color$groupLabel)
  pr = scale(prcomp(t(scale(data_obj$fpkm_for_clust)))$x)
  annotation.df$pca.1 <- pr[, 1]
  annotation.df$pca.2 <- pr[, 2]

  return(annotation.df)
}
```
```{r get_SC3, echo = TRUE}
get_SC3 <- function(eData, is_normalized =FALSE) {
  require(SingleCellExperiment)
  require(SC3)
  
  
  if(!is_normalized) {
    sce <- SingleCellExperiment(
      assays = list(
      counts = eData,
      logcounts = log2(eData + 1)
      ), colData = annotation.df) 
    } else {
      sce <- SingleCellExperiment(
        assays = list(
          counts = 2^eData,
          logcounts = eData
          ), colData = annotation.df)
      
    }
  
  rowData(sce)$feature_symbol <- rownames(sce)
  
  sce <- sc3_estimate_k(sce)
  k_est <- sce@metadata$sc3$k_estimation
  sce <- sc3(sce, ks = k_est, biology = FALSE, n_cores=2,  k_estimator = FALSE, rand_seed=2333333, gene_filter = FALSE)
  
  eval(parse(text=paste0("ans <- colData(sce)$sc3_", k_est, "_clusters")))

  annotation.df$clustering <- ans
  return(annotation.df)
}  
```
```{r get_Seurat, echo = TRUE}
get_Seurat <- function(eData, resolution = 0.6, is_normalized = FALSE) {
  
  require(Seurat)
  
  hi_var_features <- rownames(eData)
  
  srt <- CreateSeuratObject(eData, project = prefix)
  
  if(!is_normalized) {
    
    srt <- NormalizeData(srt, normalization.method = "LogNormalize", scale.factor = 10000, verbose = FALSE)
    srt <- ScaleData(srt, features = hi_var_features, verbose = FALSE)
  } else {
    
    srt@assays$RNA@scale.data <- eData
  }
  
  srt <- RunPCA(srt, features = hi_var_features, verbose = FALSE)
  
  srt <- FindNeighbors(srt, verbose = FALSE)
  srt <- FindClusters(object = srt, resolution = resolution, 
                      algorithm = 1, n.start = 100, graph.name = NULL,
                      n.iter = 10, random.seed = 0, verbose = FALSE)
  
  annotation.df$clustering = as.factor(srt@active.ident)
  
  tmp <- scale(srt@reductions$pca@cell.embeddings)
  annotation.df$pca.1 <- tmp[, 1]
  annotation.df$pca.2 <- tmp[, 2]
  
  srt <- RunUMAP(srt, features = hi_var_features, verbose = FALSE)
  tmp <- scale(srt@reductions$umap@cell.embeddings)
  annotation.df$umap.1 <- tmp[, 1]
  annotation.df$umap.2 <- tmp[, 2]
  
  srt <- RunTSNE(srt, features = hi_var_features)
  tmp <- scale(srt@reductions$tsne@cell.embeddings)
  annotation.df$tsne.1 <- tmp[, 1]
  annotation.df$tsne.2 <- tmp[, 2]
  
  return(annotation.df)
}
```
```{r, get_clusterExperiment, echo = TRUE}
get_clusterExperiment <- function(eData, is_normalized = FALSE) {
  require(clusterExperiment)
  
  if(!is_normalized) {
    sce <- SingleCellExperiment(
      assays = list(
        counts = eData,
        logcounts = log2(eData + 1)
      ), colData = annotation.df) 
  } else {
    sce <- SingleCellExperiment(
      assays = list(
        counts = 2^eData,
        logcounts = eData
      ), colData = annotation.df)
  }
  
  reducedDim(sce, "PCA") <- irlba::prcomp_irlba(assays(sce, "logcounts", n=5))$x 
  
  se = RSEC(sce, 
            whichAssay = "logcounts", 
            minSizes=5, reduceMethod="PCA", nReducedDims=5, ncores=4, random.seed=176201,
            dendroReduce="PCA", dendroNDims = 5, 
            mergeMethod = "adjP", mergeDEMethod = "limma", 
            mergeCutoff = 0.1, mergeLogFCcutoff = 1)
  
  tmp <- primaryCluster(se)
  tmp[tmp < 0] <- NA
  
  annotation.df$clustering <- tmp
  
  return(annotation.df)
}
```

# sc_10x_5cl
```{r, eval = file.exists("sc_10x_5cl_expression_working.RData"), echo=FALSE}
load("sc_10x_5cl_expression_working.RData")
annotation.df <- annotation.df[hhc$labels,]
what <- "AWST protocol"
```
```{r sc_10x_5cl, eval = !file.exists("sc_10x_5cl_expression_working.RData"), echo=TRUE}
con <- gzcon(url("https://github.com/LuyiTian/sc_mixology/blob/master/data/csv/sc_10x_5cl.metadata.csv.gz?raw=true"))
annotation.df <- read.csv(textConnection(readLines(con)))
annotation.df$cell.col <- factor(annotation.df$cell_line)
levels(annotation.df$cell.col) <- c("gold", "red", "blue", "magenta", "green3")
annotation.df$cell.col <- as.character(annotation.df$cell.col)
annotation.df$cell_line <- paste0("mix", annotation.df$mix)

con <- gzcon(url("https://github.com/LuyiTian/sc_mixology/blob/master/data/csv/sc_10x_5cl.count.csv.gz?raw=true"))
ddata <- read.csv(textConnection(readLines(con)))

both <- intersect(colnames(ddata), rownames(annotation.df))
ddata <- ddata[, both]
annotation.df <- annotation.df[both,]
prefix <- "sc_10x_5cl"
save(ddata, annotation.df, prefix, file = paste0(prefix, "_counts.RData"))
```
```{r _working.RData, eval= !file.exists("sc_10x_expression_working.RData"), echo=TRUE}
require(awst)
require(EDASeq)
require(Rtsne)
require(umap)
require(SingleCellExperiment)
####
#load(paste0(prefix, "_counts.RData"))
#load(paste0(prefix, "_expression.RData"))
####
no_of_detected_gene_per_sample <- colSums(ddata > 0) 
fivenum(no_of_detected_gene_per_sample)
ddata <- EDASeq::betweenLaneNormalization(as.matrix(ddata), which = "full", round = FALSE)
# apply the AWS-transformation 
tmp <- rowSums(ddata)
sum(tmp > 0)
tmp <- colSums(ddata)
sum(tmp > 0)

exprData <- awst(ddata, poscount = TRUE, full_quantile = TRUE)
sum(is.na(rowSums(exprData)))
sum(is.na(colSums(exprData)))

save(exprData, prefix, file = paste0(prefix, "_expression.RData"))
#dim(exprData <- gene_filter(exprData))
#write.table(exprData, file =  paste0(prefix, "_expression.tsv"), sep = "\t")

nrow_exprData <- nrow(exprData)
ncol_exprData <- ncol(exprData)
ddist <- dist(exprData)
save(ddist, nrow_exprData, ncol_exprData, prefix, file = paste0(prefix, "_expression_dist.RData"))

hhc <- hclust(ddist, method = "ward.D2")

aCalinski <- calinski(hhc)

pprcomp <- prcomp(exprData) # Run PC analysis
pprcomp$x <- pprcomp$x[, 1:5]
pprcomp$rotation <- pprcomp$rotation[, 1:5]

set.seed(2020) 
ans_Rtsne <- Rtsne(exprData, pca = FALSE) # Run TSNE

set.seed(2020) 
ans_umap <- umap(exprData) # Run Umap

require(SingleCellExperiment)

rm(tmp)
tmp <- get_RCA(ddata)
wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
ans_RCA_raw  <- tmp[, -wwhich]
#    user   system  elapsed 

rm(tmp)
system.time({tmp <- get_RCA(t(exprData), is_normalized = TRUE)})
wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
ans_RCA_awst  <- tmp[, -wwhich]
#    user   system  elapsed 

system.time({ans_SC3_raw <- get_SC3(as.matrix(ddata))[, c("cell.col", "clustering")]})
#    user   system  elapsed 
# 130.476    9.860 2821.177 sc_10x_5cl
#  12.496    2.868  474.091 sc_10x
#   2.284    0.892   55.425 RNAmix_celseq2

system.time({ans_SC3_awst <- get_SC3(t(exprData), is_normalized = TRUE)[, c("cell.col",  "clustering")]})
#    user   system  elapsed 
# 150.964   13.932 3057.959 sc_10x_5cl
#  12.664    2.572  466.772 sc_10x
#   2.540    0.888   57.347 RNAmix_celseq2

system.time({ans_clusterExp_raw <- get_clusterExperiment(ddata, is_normalized = FALSE)[, c("cell.col",  "clustering")]})

system.time({ans_clusterExp_awst <- get_clusterExperiment(t(exprData), is_normalized = TRUE)[, c("cell.col",  "clustering")]})

rm(tmp)
system.time({tmp <- get_Seurat(ddata, is_normalized = FALSE)})
wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
ans_Seurat_raw_LoR  <- tmp[, -wwhich]
#   user  system elapsed 
# 14.412   0.432  14.902 RNAmix_celseq2
# 36.192   0.832  37.031 sc_10x

rm(tmp)
system.time({tmp <- get_Seurat(t(exprData), is_normalized = TRUE)})
wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
ans_Seurat_awst_LoR  <- tmp[, -wwhich]
#   user  system elapsed 
# 11.464   0.184  11.651 RNAmix_celseq2
# 33.304   1.124  34.431 sc_10x

rm(tmp)
system.time({tmp <- get_Seurat(ddata, is_normalized = FALSE, resolution = 1.6)})
wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
ans_Seurat_raw_HiR  <- tmp[, -wwhich]
#   user  system elapsed 
# 11.560   0.152  11.713 RNAmix_celseq2
# 35.472   0.792  36.280 sc_10x

rm(tmp)
system.time({tmp <- get_Seurat(t(exprData), is_normalized = TRUE, resolution = 1.6)})
wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
ans_Seurat_awst_HiR  <- tmp[, -wwhich]
#   user  system elapsed 
# 11.692   0.140  11.836 RNAmix_celseq2
# 34.184   1.232  35.418 sc_10x

rm(tmp)
system.time({tmp <- get_RaceID(ddata, is_normalized = FALSE)})
wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
ans_RaceID_raw  <- tmp[, -wwhich]
#   user  system elapsed 
#  7.004   0.116   7.131 RNAmix_celseq2
# 41.932   0.532  42.429 sc_10x
  
rm(tmp)
system.time({tmp <- get_RaceID(2^t(exprData), is_normalized = TRUE)})
wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
ans_RaceID_awst  <- tmp[, -wwhich]
#   user  system elapsed 
#  5.784   0.092   5.866 RNAmix_celseq2
# 47.252   0.560  47.759 sc_10x

### VST  
if(!file.exists(paste0(prefix, "_VST.RData"))) {
  load(paste0(prefix, "_counts.RData"))
  require(DESeq2)
  dds <- DESeqDataSetFromMatrix(countData = as.matrix(ddata), 
                              colData = annotation.df, 
                              design = ~1)
# vvst <- vst(dds, blind=FALSE)
#Error in vst(dds, blind = FALSE) : 
#  less than 'nsub' rows with mean normalized count > 5, 
#  it is recommended to use varianceStabilizingTransformation directly

  system.time({vvst <- varianceStabilizingTransformation(dds, blind=FALSE)})
#    user   system  elapsed 
#1426.620    2.388 1430.940 sc_10x
# 157.372    0.020  157.395 RNAmix
  
  vstData <- assay(vvst)
  save(vstData, file = paste0(prefix, "_VST.RData"))
} else load(paste0(prefix, "_VST.RData"))
  
  rm(tmp)
  system.time({tmp <- get_Seurat(vstData, is_normalized = TRUE)})
#   user  system elapsed 
#267.628   5.892 327.423 sc_10x_5cl
# 10.652   0.112  10.765 RNAmix_celseq2
# 33.912   0.584  34.494 sc_10x
  wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
  ans_Seurat_VST_LoR  <- tmp[, -wwhich]
  
  rm(tmp)
  system.time({tmp <- get_Seurat(vstData, is_normalized = TRUE, resolution = 1.6)})
  wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
  ans_Seurat_VST_HiR  <- tmp[, -wwhich]
#   user  system elapsed 
#286.520   5.096 321.105 sc_10x_5cl
# 11.080   0.128  11.212 RNAmix_celseq2
# 34.228   0.620  34.858 sc_10x
  
  rm(tmp)
  system.time({tmp <- get_RaceID(2^vstData, is_normalized = TRUE)})
  wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
  ans_RaceID_VST  <- tmp[, -wwhich]
#    user   system  elapsed 
#2743.628    4.196 2777.134 sc_10x_5cl
#   8.992    0.096    9.055 RNAmix_celseq2
#  51.904    0.744   52.610 sc_10x

  
system.time({ans_SC3_VST <- get_SC3(vstData, is_normalized = TRUE)[, c("cell.col",  "clustering")]})
#    user   system  elapsed 
# 145.884   10.376 2626.021 sc_10x_5cl
#  13.448    2.224  459.802 sc_10x
#   2.460    0.836   53.050 RNAmix_celseq2

system.time({ans_clusterExp_VST <- get_clusterExperiment(vstData, is_normalized = TRUE)[, c("cell.col",  "clustering")]})
  
rm(tmp)
system.time({tmp <- get_RCA(vstData, is_normalized = TRUE)})
wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
ans_RCA_vst  <- tmp[, -wwhich]  
#   user  system elapsed 
#  8.208   0.000  40.342 
  
  
if(!file.exists(paste0(prefix, "_scry.RData"))) {
  load(paste0(prefix, "_counts.RData"))
  
  library(scry)
  scryData_pearson <- nullResiduals(as.matrix(ddata), type = "pearson")
  save(scryData_pearson, file = paste0(prefix, "_scry.RData"))
  
} else load(paste0(prefix, "_scry.RData"))
  
  rm(tmp)
  system.time({tmp <- get_Seurat(scryData_pearson, is_normalized = TRUE)})
  wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
  ans_Seurat_scry_LoR  <- tmp[, -wwhich]
#   user  system elapsed 
#284.504   8.032 356.628 sc_10x_5cl
# 16.140   0.172  16.313 RNAmix_celseq2
# 34.216   0.724  34.948 sc_10x
  
  rm(tmp)
  system.time({tmp <- get_Seurat(scryData_pearson, is_normalized = TRUE, resolution = 1.6)})
  wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
  ans_Seurat_scry_HiR  <- tmp[, -wwhich]
#   user  system elapsed 
#293.972   2.968 320.115 sc_10x_5cl 
# 15.428   0.192  15.620 RNAmix_celseq2
# 34.496   0.908  35.406 sc_10x
  
  
  rm(tmp)
  system.time({tmp <- get_RaceID(2^scryData_pearson, is_normalized = TRUE)})
  wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
  ans_RaceID_scry  <- tmp[, -wwhich]
#    user   system  elapsed 
#2387.520    2.580 2407.494 sc_10x_5cl
#   6.892    0.100    6.976 RNAmix_celseq2
#  59.864    0.528   60.362 sc_10x

  rm(tmp)
  system.time({tmp <- get_RCA(scryData_pearson, is_normalized = TRUE)})
  wwhich <- which(colnames(tmp) %in% colnames(annotation.df))
  ans_RCA_scry  <- tmp[, -wwhich]
#    user   system  elapsed 
#2387.520    2.580 2407.494 sc_10x_5cl
#   7.920    0.000  309.245 sc_10x
  
system.time({ans_SC3_scry <- get_SC3(scryData_pearson, is_normalized = TRUE)[, c("cell.col",  "clustering")]})
#    user   system  elapsed 
# 143.712   11.704 4087.033 sc_10x_5cl
#  14.312    3.232 1097.622 sc_10x
#   2.484    0.848   76.398 RNAmix_celseq2

system.time({ans_clusterExp_scry <- get_clusterExperiment(scryData, is_normalized = TRUE)[, c("cell.col",  "clustering")]})

if(file.exists(paste0(prefix, "_rsec_working.RData"))) { 
  load(paste0(prefix, "_rsec_working.RData"))
  
  rm(tmp)
  tmp <- annotation.df[attr(ans_clusterExp_raw, "rownames"),]
  tmp$clustering <- attr(ans_clusterExp_raw, "listData")$clustering_res
  tmp$clustering[tmp$clustering < 0] <- NA
  tmp$clustering <- factor(tmp$clustering)
  ans_clusterExp_raw <- tmp[, c("cell.col", "clustering")]

  rm(tmp)
  tmp <- annotation.df[attr(ans_clusterExp_awst, "rownames"),]
  tmp$clustering <- attr(ans_clusterExp_awst, "listData")$clustering_res
  tmp$clustering[tmp$clustering < 0] <- NA
  tmp$clustering <- factor(tmp$clustering)
  ans_clusterExp_awst <- tmp[, c("cell.col", "clustering")]

  rm(tmp)
  tmp <- annotation.df[attr(ans_clusterExp_scry, "rownames"),]
  tmp$clustering <- attr(ans_clusterExp_scry, "listData")$clustering_res
  tmp$clustering[tmp$clustering < 0] <- NA
  tmp$clustering <- factor(tmp$clustering)
  ans_clusterExp_scry <- tmp[, c("cell.col", "clustering")]

  tmp <- annotation.df[attr(ans_clusterExp_vst, "rownames"),]
  tmp$clustering <- attr(ans_clusterExp_vst, "listData")$clustering_res
  tmp$clustering[tmp$clustering < 0] <- NA
  tmp$clustering <- factor(tmp$clustering)
  ans_clusterExp_vst <- tmp[, c("cell.col", "clustering")]
}
  
  
  
  
save(hhc, nrow_exprData, ncol_exprData, annotation.df, aCalinski,
     pprcomp, ans_Rtsne, ans_umap, prefix, 
     ans_RaceID_awst, ans_RaceID_raw,
     ans_RaceID_scry, 
     ans_RaceID_VST,
     ans_RCA_raw, ans_RCA_awst, 
     ans_RCA_vst, ans_RCA_scry,
     ans_SC3_raw, ans_SC3_awst, 
     ans_SC3_VST, ans_SC3_scry, 
     ans_Seurat_raw_LoR, ans_Seurat_raw_HiR,
     ans_Seurat_awst_LoR, ans_Seurat_awst_HiR,
     ans_Seurat_VST_LoR, ans_Seurat_VST_HiR,
     ans_Seurat_scry_LoR, ans_Seurat_scry_HiR,
     ans_clusterExp_raw, ans_clusterExp_awst,
     ans_clusterExp_scry, ans_clusterExp_vst, 
     file = paste0(prefix, "_expression_working.RData"))
```
```{r _main_dendrogram.pdf, echo=FALSE, eval=TRUE, fig.width=30, fig.height=13, echo=FALSE, eval=TRUE, results='asis'}
cat("\n\n### Dendrogram of ", prefix, "\n")
clustering.df <- data.frame(cell_line = annotation.df$cell_line, 
                            True.col = annotation.df$cell.col,
                            row.names = rownames(annotation.df))
if("mRNA_amount.col" %in% colnames(annotation.df)) clustering.df$mRNA_amount.col <-  annotation.df$mRNA_amount.col
###
mmain  <-  paste0(prefix, " (", nrow_exprData, " cells/", ncol_exprData, " genes)")
if(save_plots) {
  mmain <- ""
  png(paste0(prefix, "_expression_dist_hclust.png"), 
      width= png_width_large, height= png_height_large, res = png_res)
}
hhc$height <- hhc$height/max(hhc$height)
plot(hhc, hang = -1, labels = FALSE, xlab = "", sub = "", main = mmain)
add_Calinski_curve(from = floor(.1*length(hhc$labels)), to = floor(.9*length(hhc$labels)))

wwhere <- which(aCalinski[2:length(aCalinski)] - aCalinski[1:(length(aCalinski)-1)] < 0)[1]

hh <- mean(c(hhc$height[length(hhc$height)-wwhere+2], hhc$height[length(hhc$height)-wwhere+1]))
abline(h = hh, col = "red")
clustering.df$estimate.col <- as.factor(cutree(hhc, k = wwhere))
tt <- table(clustering.df$estimate.col, clustering.df$True.col)
colorCode <- colnames(tt)[apply(tt, 1, which.max)]
if(prefix == "RNAmix_celseq2") colorCode <- palette()[1:7]
levels(clustering.df$estimate.col) <- colorCode

clustering.col <- clustering.df[, grep(".col", colnames(clustering.df))]
colnames(clustering.col) <- gsub(".col", "", colnames(clustering.col))
colored_bars(colors = clustering.col, dend = as.dendrogram(hhc), y_scale = .15, y_shift = 0)

tt <- table(clustering.df$cell_line, clustering.df$True.col)
tmp <- rowSums(tt)
pct <- paste0(round(100*tt/sum(tt), 1), "%")
llegend <- paste(names(tmp), " (", tmp, ")", sep = "")
ffill <- colnames(tt)[apply(tt, 1, which.max)]
legend(floor(.7*length(hhc$labels)), 1, legend=llegend, fill = ffill, y.intersp = 1, box.col = "white", border = "white", title = prefix, title.adj = 0)

results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$estimate.col))

clustering.df$cell.col <- as.character(clustering.df$True.col) #??????
```
```{r PCA, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### PCA of ", prefix, "\n")
x_legend <- 0.08; y_legend <- -1.75
x_text <- -2.3; y_text <- 2.
pprcomp$x <- scale(pprcomp$x)
#
mmain = paste0("Principal components analysis (", prefix, "; ", nrow_exprData, " cells/", ncol_exprData, " features)")
#if(save_plots) png(paste0(jobName, "_expression_prcomp_CITE6.png"), width= png_width_small, height= png_height_small, res = png_res)
plot(pprcomp$x, col = annotation.df$cell.col, main = mmain, 
     xlab = "first principal component", ylab = "second principal component", pch = 19)
```
```{r Rtsne, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### t-SNE of ", prefix, "\n")
ans_Rtsne$Y <- scale(ans_Rtsne$Y)
#if(save_plots) png(paste0(jobName, "_expression_Rtsne.png"), width= png_width_small, height= png_height_small, res = png_res)
mmain = "T-distributed Stochastic Neighbor Embedding (tsne)"
plot(ans_Rtsne$Y[, 1], ans_Rtsne$Y[, 2], col = annotation.df$cell.col, main = mmain, xlab = "", ylab = "", pch = 19)
```
```{r umap, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### umap of ", prefix, "\n")
ans_umap$layout <- scale(ans_umap$layout)
#if(save_plots) png(paste0(jobName, "_expression_umap.png"), width= png_width_small, height= png_height_small, res = png_res)
mmain = "Uniform Manifold Approximation and Projection (umap)"
plot(ans_umap$layout[, 1], ans_umap$layout[, 2], col = annotation.df$cell.col, main = mmain, xlab = "", ylab = "", pch = 19)
```

```{r raceID_raw, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### counts + RaceID (default + umap) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_RaceID_raw),]
clustering.df$RaceID_raw <- clustering.df$tmp <- ans_RaceID_raw$clustering
what <- "RaceID + counts"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_RaceID_raw$umap.1, ans_RaceID_raw$umap.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "")

#cat("\n\n### AWST + RaceID (default + t-SNE) of ", prefix, "\n")
#plot(ans_RaceID_raw$tsne.1, ans_RaceID_raw$tsne.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "")
```
```{r raceID_awst, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### AWST + RaceID (default + umap) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_RaceID_awst),]
clustering.df$RaceID_awst <- clustering.df$tmp <- ans_RaceID_awst$clustering
what <- "RaceID + AWST"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_RaceID_awst$umap.1, ans_RaceID_awst$umap.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "")

#cat("\n\n### AWST + RaceID (default + t-SNE) of ", prefix, "\n")
#plot(ans_RaceID_awst$tsne.1, ans_RaceID_awst$tsne.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "")
```
```{r raceID_vst, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### VST + RaceID (default + umap) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_RaceID_awst),]
clustering.df$RaceID_vst <- clustering.df$tmp <- ans_RaceID_VST$clustering
what <- "RaceID + VST"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_RaceID_VST$umap.1, ans_RaceID_VST$umap.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "")

#cat("\n\n### VST + RaceID (default + t-SNE) of ", prefix, "\n")
#plot(ans_RaceID_VST$tsne.1, ans_RaceID_VST$tsne.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "")
```
```{r raceID_scry, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### Townes + RaceID (default + umap) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_RaceID_scry),]
clustering.df$RaceID_scry <- clustering.df$tmp <- ans_RaceID_scry$clustering
what <- "RaceID + Townes"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_RaceID_scry$umap.1, ans_RaceID_scry$umap.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "")

#cat("\n\n### Townes + RaceID (default + t-SNE) of ", prefix, "\n")
#plot(ans_RaceID_scry$tsne.1, ans_RaceID_scry$tsne.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "")
```

```{r RCA_raw, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### counts + RCA (pca) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_RCA_raw),]
clustering.df$RCA_raw <- clustering.df$tmp <- ans_RCA_raw$clustering
what <- "RCA + counts"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_RCA_raw$pca.1, ans_RCA_raw$pca.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "")
```
```{r RCA_awst, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### AWST + RCA (pca) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_RCA_awst),]
clustering.df$RCA_awst <- clustering.df$tmp <- ans_RCA_awst$clustering
what <- "RCA +  AWST"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_RCA_awst$pca.1, ans_RCA_awst$pca.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "")
```
```{r RCA_vst, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### vst + RCA (pca) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_RCA_raw),]
clustering.df$RCA_raw <- clustering.df$tmp <- ans_RCA_raw$clustering
what <- "RCA + VST"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_RCA_raw$pca.1, ans_RCA_raw$pca.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "")
```
```{r RCA_scry, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### Townes + RCA (pca) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_RCA_scry),]
clustering.df$RCA_raw <- clustering.df$tmp <- ans_RCA_scry$clustering
what <- "RCA + Townes"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_RCA_scry$pca.1, ans_RCA_scry$pca.2, 
#    col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "")
```

```{r Seurat_raw_LoR, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### counts + Seurat  (low resolution, pca) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_Seurat_raw_LoR),]
clustering.df$Seurat_raw_LoR <- clustering.df$tmp <- ans_Seurat_raw_LoR$clustering
what <- "Seurat (LoRes) + counts"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_Seurat_raw_LoR$pca.1, ans_Seurat_raw_LoR$pca.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

clustering.df$tmp <- factor(clustering.df$tmp)
levels(clustering.df$tmp) <- ccolors(length(levels(clustering.df$tmp)))
clustering.df$tmp <- as.character(clustering.df$tmp)
#plot(ans_Seurat_raw_LoR$pca.1, ans_Seurat_raw_LoR$pca.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### counts + Seurat  (low resolution, umap) of ", prefix, "\n")
#plot(ans_Seurat_raw_LoR$umap.1, ans_Seurat_raw_LoR$umap.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_raw_LoR$umap.1, ans_Seurat_raw_LoR$umap.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### counts + Seurat  (low resolution, tsne) of ", prefix, "\n")
#plot(ans_Seurat_raw_LoR$tsne.1, ans_Seurat_raw_LoR$tsne.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_raw_LoR$tsne.1, ans_Seurat_raw_LoR$tsne.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")
```
```{r Seurat_awst_LoR, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### awst + Seurat  (low resolution, pca) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_Seurat_awst_LoR),]
clustering.df$Seurat_awst_LoR <- clustering.df$tmp <- ans_Seurat_awst_LoR$clustering
what <- "Seurat (LoRes) +  AWST"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_Seurat_awst_LoR$pca.1, ans_Seurat_awst_LoR$pca.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#clustering.df$tmp <- factor(clustering.df$tmp)
#levels(clustering.df$tmp) <- ccolors(length(levels(clustering.df$tmp)))
#clustering.df$tmp <- as.character(clustering.df$tmp)
#plot(ans_Seurat_awst_LoR$pca.1, ans_Seurat_awst_LoR$pca.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### awst + Seurat  (low resolution, umap) of ", prefix, "\n")
#plot(ans_Seurat_awst_LoR$umap.1, ans_Seurat_awst_LoR$umap.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_awst_LoR$umap.1, ans_Seurat_awst_LoR$umap.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### awst + Seurat  (low resolution, tsne) of ", prefix, "\n")
#plot(ans_Seurat_awst_LoR$tsne.1, ans_Seurat_awst_LoR$tsne.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_awst_LoR$tsne.1, ans_Seurat_awst_LoR$tsne.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")
```
```{r Seurat_VST_LoR, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### VST + Seurat  (low resolution, pca) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_Seurat_VST_LoR),]
clustering.df$Seurat_VST_LoR <- clustering.df$tmp <- ans_Seurat_VST_LoR$clustering
what <- "Seurat (LoRes) + VST"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_Seurat_VST_LoR$pca.1, ans_Seurat_VST_LoR$pca.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#clustering.df$tmp <- factor(clustering.df$tmp)
#levels(clustering.df$tmp) <- ccolors(length(levels(clustering.df$tmp)))
#clustering.df$tmp <- as.character(clustering.df$tmp)
#plot(ans_Seurat_VST_LoR$pca.1, ans_Seurat_VST_LoR$pca.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### VST + Seurat  (low resolution, umap) of ", prefix, "\n")
#plot(ans_Seurat_VST_LoR$umap.1, ans_Seurat_VST_LoR$umap.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_VST_LoR$umap.1, ans_Seurat_VST_LoR$umap.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### VST + Seurat  (low resolution, tsne) of ", prefix, "\n")
#plot(ans_Seurat_VST_LoR$tsne.1, ans_Seurat_VST_LoR$tsne.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_VST_LoR$tsne.1, ans_Seurat_VST_LoR$tsne.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")
```
```{r Seurat_scry_LoR, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### Townes + Seurat  (low resolution, pca) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_Seurat_scry_LoR),]
clustering.df$Seurat_scry_LoR <- clustering.df$tmp <- ans_Seurat_scry_LoR$clustering
what <- "Seurat (LoRes) + Townes"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_Seurat_scry_LoR$pca.1, ans_Seurat_scry_LoR$pca.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#clustering.df$tmp <- factor(clustering.df$tmp)
#levels(clustering.df$tmp) <- ccolors(length(levels(clustering.df$tmp)))
#clustering.df$tmp <- as.character(clustering.df$tmp)
#plot(ans_Seurat_scry_LoR$pca.1, ans_Seurat_scry_LoR$pca.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### Townes + Seurat  (low resolution, umap) of ", prefix, "\n")
#plot(ans_Seurat_scry_LoR$umap.1, ans_Seurat_scry_LoR$umap.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_scry_LoR$umap.1, ans_Seurat_scry_LoR$umap.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### Townes + Seurat  (low resolution, tsne) of ", prefix, "\n")
#plot(ans_Seurat_scry_LoR$tsne.1, ans_Seurat_scry_LoR$tsne.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_scry_LoR$tsne.1, ans_Seurat_scry_LoR$tsne.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")
```
```{r Seurat_raw_HiR, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### counts + Seurat  (high resolution, pca) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_Seurat_raw_HiR),]
clustering.df$Seurat_raw_HiR <- clustering.df$tmp <- ans_Seurat_raw_HiR$clustering
what <- "Seurat (HiRes) +  counts"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_Seurat_raw_HiR$pca.1, ans_Seurat_raw_HiR$pca.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#clustering.df$tmp <- factor(clustering.df$tmp)
#levels(clustering.df$tmp) <- ccolors(length(levels(clustering.df$tmp)))
#clustering.df$tmp <- as.character(clustering.df$tmp)
#plot(ans_Seurat_raw_HiR$pca.1, ans_Seurat_raw_HiR$pca.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### counts + Seurat  (high resolution, umap) of ", prefix, "\n")
#plot(ans_Seurat_raw_HiR$umap.1, ans_Seurat_raw_HiR$umap.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_raw_HiR$umap.1, ans_Seurat_raw_HiR$umap.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### counts + Seurat  (high resolution, tsne) of ", prefix, "\n")
#plot(ans_Seurat_raw_HiR$tsne.1, ans_Seurat_raw_HiR$tsne.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_raw_HiR$tsne.1, ans_Seurat_raw_HiR$tsne.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")
```
```{r Seurat_awst_HiR, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### awst + Seurat  (high resolution, pca) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_Seurat_awst_HiR),]
clustering.df$Seurat_awst_HiR <- clustering.df$tmp <- ans_Seurat_awst_HiR$clustering
what <- "Seurat (HiRes) +  AWST"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_Seurat_awst_HiR$pca.1, ans_Seurat_awst_HiR$pca.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#clustering.df$tmp <- factor(clustering.df$tmp)
#levels(clustering.df$tmp) <- ccolors(length(levels(clustering.df$tmp)))
#clustering.df$tmp <- as.character(clustering.df$tmp)
#plot(ans_Seurat_awst_HiR$pca.1, ans_Seurat_awst_HiR$pca.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### awst + Seurat  (high resolution, umap) of ", prefix, "\n")
#plot(ans_Seurat_awst_HiR$umap.1, ans_Seurat_awst_HiR$umap.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_awst_HiR$umap.1, ans_Seurat_awst_HiR$umap.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### awst + Seurat  (high resolution, tsne) of ", prefix, "\n")
#plot(ans_Seurat_awst_HiR$tsne.1, ans_Seurat_awst_HiR$tsne.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_awst_HiR$tsne.1, ans_Seurat_awst_HiR$tsne.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")
```
```{r Seurat_VST_HiR, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### VST + Seurat  (high resolution, pca) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_Seurat_VST_HiR),]
clustering.df$Seurat_VST_HiR <- clustering.df$tmp <- ans_Seurat_VST_HiR$clustering
what <- "Seurat (HiRes) + VST"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_Seurat_VST_HiR$pca.1, ans_Seurat_VST_HiR$pca.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#clustering.df$tmp <- factor(clustering.df$tmp)
#levels(clustering.df$tmp) <- ccolors(length(levels(clustering.df$tmp)))
#clustering.df$tmp <- as.character(clustering.df$tmp)
#plot(ans_Seurat_VST_HiR$pca.1, ans_Seurat_VST_HiR$pca.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### VST + Seurat  (high resolution, umap) of ", prefix, "\n")
#plot(ans_Seurat_VST_HiR$umap.1, ans_Seurat_VST_HiR$umap.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_VST_HiR$umap.1, ans_Seurat_VST_HiR$umap.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### VST + Seurat  (high resolution, tsne) of ", prefix, "\n")
#plot(ans_Seurat_VST_HiR$tsne.1, ans_Seurat_VST_HiR$tsne.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_VST_HiR$tsne.1, ans_Seurat_VST_HiR$tsne.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")
```
```{r Seurat_scry_HiR, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### Townes + Seurat  (high resolution, pca) of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_Seurat_scry_HiR),]
clustering.df$Seurat_scry_LoR <- clustering.df$tmp <- ans_Seurat_scry_HiR$clustering
what <- "Seurat (HiRes) + Townes"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

#plot(ans_Seurat_scry_HiR$pca.1, ans_Seurat_scry_HiR$pca.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#clustering.df$tmp <- factor(clustering.df$tmp)
#levels(clustering.df$tmp) <- ccolors(length(levels(clustering.df$tmp)))
#clustering.df$tmp <- as.character(clustering.df$tmp)
#plot(ans_Seurat_scry_HiR$pca.1, ans_Seurat_scry_HiR$pca.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### Townes + Seurat  (high resolution, umap) of ", prefix, "\n")
#plot(ans_Seurat_scry_HiR$umap.1, ans_Seurat_scry_HiR$umap.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_scry_HiR$umap.1, ans_Seurat_scry_HiR$umap.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")

#cat("\n\n### Townes + Seurat  (high resolution, tsne) of ", prefix, "\n")
#plot(ans_Seurat_scry_HiR$tsne.1, ans_Seurat_scry_HiR$tsne.2, 
#     col= clustering.df$cell.col, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with theoretical partition")

#plot(ans_Seurat_scry_HiR$tsne.1, ans_Seurat_scry_HiR$tsne.2, 
#     col= clustering.df$tmp, pch = 19, xlab = "", ylab = "", 
#     main = "Cells annotated with empirical partition")
```

```{r SC3, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### counts + SC3 of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_SC3_raw),]
clustering.df$SC3_raw <- clustering.df$tmp <- ans_SC3_raw$clustering
what <- "SC3 + counts"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

cat("\n\n### awst + SC3 of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_SC3_awst),]
clustering.df$SC3_awst <- clustering.df$tmp <- ans_SC3_awst$clustering
what <- "SC3 + AWST"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

cat("\n\n### VST + SC3 of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_SC3_VST),]
clustering.df$SC3_vst <- clustering.df$tmp <- ans_SC3_VST$clustering
what <- "SC3 + VST"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))

cat("\n\n### Townes + SC3 of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_SC3_scry),]
clustering.df$SC3_scry <- clustering.df$tmp <- ans_SC3_scry$clustering
what <- "SC3 + Townes"
results[k <- k+1,] <- c(prefix, what, get_metrics(clustering.df$True.col, clustering.df$tmp))
```

```{r clustExp, eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
cat("\n\n### counts + clustExp of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_clusterExp_raw),]
clustering.df$clustExp_raw <- clustering.df$tmp <- ans_clusterExp_raw$clustering
tmp.df <- data.frame(theo = clustering.df$True.col, emp = clustering.df$tmp)
tmp.df <- tmp.df[rowSums(0 + is.na(tmp.df)) == 0,]
what <- "clustExp + counts"
results[k <- k+1,] <- c(prefix, what, get_metrics(tmp.df$theo, tmp.df$emp))

cat("\n\n### awst + clustExp of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_clusterExp_awst),]
clustering.df$clusterExp_awst <- clustering.df$tmp <- ans_clusterExp_awst$clustering
tmp.df <- data.frame(theo = clustering.df$True.col, emp = clustering.df$tmp)
tmp.df <- tmp.df[rowSums(0 + is.na(tmp.df)) == 0,]
what <- "clustExp + AWST"
results[k <- k+1,] <- c(prefix, what, get_metrics(tmp.df$theo, tmp.df$emp))

cat("\n\n### VST + clustExp of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_clusterExp_vst),]
clustering.df$clusterExp_vst <- clustering.df$tmp <- ans_clusterExp_vst$clustering
tmp.df <- data.frame(theo = clustering.df$True.col, emp = clustering.df$tmp)
tmp.df <- tmp.df[rowSums(0 + is.na(tmp.df)) == 0,]
what <- "clustExp + VST"
results[k <- k+1,] <- c(prefix, what, get_metrics(tmp.df$theo, tmp.df$emp))

cat("\n\n### Townes + clustExp of ", prefix, "\n")
clustering.df <- clustering.df[rownames(ans_clusterExp_scry),]
clustering.df$clusterExp_scry <- clustering.df$tmp <- ans_clusterExp_scry$clustering
tmp.df <- data.frame(theo = clustering.df$True.col, emp = clustering.df$tmp)
tmp.df <- tmp.df[rowSums(0 + is.na(tmp.df)) == 0,]
what <- "clustExp + Townes"
results[k <- k+1,] <- c(prefix, what, get_metrics(tmp.df$theo, tmp.df$emp))
```


# sc_10x
```{r, eval = file.exists("sc_10x_expression_working.RData"), echo=FALSE}
load("sc_10x_expression_working.RData")
annotation.df <- annotation.df[hhc$labels,]
what <- "AWST protocol"
```
```{r sc_10x, eval = !file.exists("sc_10x_expression_working.RData"), echo=FALSE}
con <- gzcon(url("https://github.com/LuyiTian/sc_mixology/blob/master/data/csv/sc_10x.metadata.csv.gz?raw=true"))
annotation.df <- read.csv(textConnection(readLines(con)))
annotation.df$cell.col <- factor(annotation.df$cell_line)
levels(annotation.df$cell.col) <- c("red", "blue", "green3")
annotation.df$cell.col <- as.character(annotation.df$cell.col)

con <- gzcon(url("https://github.com/LuyiTian/sc_mixology/blob/master/data/csv/sc_10x.count.csv.gz?raw=true"))
ddata <- read.csv(textConnection(readLines(con)))

both <- intersect(colnames(ddata), rownames(annotation.df))
ddata <- ddata[, both]
annotation.df <- annotation.df[both,]
prefix <- "sc_10x"
save(ddata, annotation.df, prefix, file = paste0(prefix, "_counts.RData"))
```
```{r, ref.label= "_working.RData", eval= !file.exists("sc_10x_expression_working.RData"), echo=FALSE}
```
```{r, ref.label="_main_dendrogram.pdf", echo=FALSE, eval=TRUE, fig.width=30, fig.height=13, echo=FALSE, eval=TRUE, results='asis'}
```
```{r, ref.label="PCA", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="Rtsne", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="umap", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="raceID_raw", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="raceID_awst", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="raceID_vst", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="raceID_scry", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="RCA_raw", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="RCA_awst", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="RCA_vst", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="RCA_scry", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_raw_LoR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_awst_LoR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_VST_LoR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_scry_LoR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_raw_HiR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_awst_HiR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_VST_HiR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_scry_HiR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="SC3", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="clustExp", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```

# RNAmix_celseq2
```{r, eval = file.exists("RNAmix_celseq2_expression_working.RData"), echo=FALSE}
load("RNAmix_celseq2_expression_working.RData")
annotation.df <- annotation.df[hhc$labels,]
what <- "AWST protocol"
```
```{r RNAmix_celseq2, eval = !file.exists("RNAmix_celseq2_expression_working.RData"), echo=FALSE}
con <- gzcon(url("https://github.com/LuyiTian/sc_mixology/blob/master/data/csv/RNAmix_celseq2.metadata.csv.gz?raw=true"))
annotation.df <- read.csv(textConnection(readLines(con)))
annotation.df$cell_line <- factor(annotation.df$mix)
levels(annotation.df$cell_line)[1:2] <- "1-2"
levels(annotation.df$cell_line) <- paste("mix", levels(annotation.df$cell_line))

annotation.df$cell.col <- annotation.df$cell_line
table(annotation.df$cell.col)
levels(annotation.df$cell.col) <- c("gray25", "green4", "cyan4", "red4", "green", "cyan", "red")
annotation.df$cell.col <- as.character(annotation.df$cell.col)

annotation.df$mRNA_amount.col <- factor(annotation.df$mRNA_amount)
levels(annotation.df$mRNA_amount.col) <- c("green1", "green2", "green3", "green4")
annotation.df$mRNA_amount.col <- as.character(annotation.df$mRNA_amount.col)

con <- gzcon(url("https://github.com/LuyiTian/sc_mixology/blob/master/data/csv/RNAmix_celseq2.count.csv.gz?raw=true"))
ddata <- read.csv(textConnection(readLines(con)))

both <- intersect(colnames(ddata), rownames(annotation.df))
ddata <- ddata[, both]
annotation.df <- annotation.df[both,]
prefix <- "RNAmix_celseq2"
save(ddata, annotation.df, prefix, file = paste0(prefix, "_counts.RData"))
```
```{r, ref.label= "_working.RData", eval= !file.exists("RNAmix_celseq2_expression_working.RData"), echo=FALSE}
```
```{r, ref.label="_main_dendrogram.pdf", echo=FALSE, eval=TRUE, fig.width=30, fig.height=13, echo=FALSE, eval=TRUE, results='asis'}
```
```{r, ref.label="PCA", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="Rtsne", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="umap", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="raceID_raw", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="raceID_awst", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="raceID_vst", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="raceID_scry", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="RCA_raw", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="RCA_awst", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="RCA_vst", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="RCA_scry", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_raw_LoR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_awst_LoR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_VST_LoR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_scry_LoR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_raw_HiR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_awst_HiR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_VST_HiR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label= "Seurat_scry_HiR", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="SC3", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```
```{r, ref.label="clustExp", eval=TRUE, fig.height = 20, fig.width = 20, fig.align = "center", results='asis', echo=FALSE}
```



# Supplementary Table 3
```{r, echo=FALSE}
results <- as.data.frame(results)
results <- results[!is.na(results$ari),]
results$ari <- as.numeric(as.character(results$ari))
results$eca <- as.numeric(as.character(results$eca))
results$ecp <- as.numeric(as.character(results$ecp))
results$G <- as.numeric(as.character(results$G))

results$eca <- round(results$eca, 4)
results$ecp <- round(results$ecp, 4)
results$ari <- round(results$ari, 4)
results$G <- round(as.numeric(as.character(results$G)), 4)

knitr::kable(results)

write.table(results, file = paste0(jobName, "_results.tsv"), sep = "\t", quote = FALSE, row.names = FALSE)
```

```{r supp table 3, eval = FALSE}
rm(list = ls())
setwd("~/Dropbox/AWST/mixology/")
ttable <- read.csv("mixoloy20200925_results.tsv", sep = "\t")
jobName <- "mixoloy20200925"
library(xtable)
x.table <- xtable(ttable)
#print(xtable(ttable), include.rownames = FALSE)
```

# Figure 4
```{r _final_performance.png}
#rm(list = ls())
#setwd("~/Dropbox/AWST/mixology/")
#ttable <- read.csv("mixoloy20200925_results.tsv", sep = "\t")
#jobName <- "mixoloy20200925"
save_png <- FALSE
ttable <- results

ttable$method <- NA

awst_protocol <- ttable[grep("protocol", ttable$what),]
ttable <- ttable[-grep("protocol", ttable$what),]

ttable <- ttable[-grep("LoRes", ttable$what),]

ttable$method[grep("AWST", ttable$what)] <- "AWST"
ttable$method[grep("VST", ttable$what)] <- "VST"
ttable$method[grep("counts", ttable$what)] <- "counts"
ttable$method[grep("Townes", ttable$what)] <- "Townes"

if(save_png) png(file = paste0(jobName, "_final_performance.png"), width = 600, height = 350, res =  1/300)
boxplot(ttable$G ~ ttable$method, ylim = c(0, 1),
        xlab = "pre-processing methods",
        ylab = "performance (G index)")#, border = "gray50")
points(rep(1.13, 3), awst_protocol$G, pch = 20)
#text(rep(1.13, 3), awst_protocol$G + c(0.015, -0.055, -0.02), awst_protocol$where, pos = 4)
#text(rep(1.13, 3), awst_protocol$G + c(0.015, -0.055, -0.02), awst_protocol$what, pos = 2)
#dev.off()
```




# Session info
```{r}
sessionInfo()
```
